---
  title: "refdb_getting_data_from_GB_BOLD"
output: html_document
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## INTRODUCTION
# Creating reference DNA barcode databases

Developing and maintaining a DNA reference database can be difficult. The refdb package has been developed to assist with the different steps of the process of developing, maintaining and using DNA reference databases efficiently and reproducible. 

This document contains the scripts for downloading and combining of DNA barcode data from the NCBI GenBank and BOLD databases and the enables this data to be combined with private unpublished data (i.e., your own sequence data). It primarily relies on the refdb package except for downloading data from NCBI GenBank and for primer trimming to create amplicon specific libraries which can be paired to metabarcoding amplicons which is completed using cutadapt (in python)). There are also a few bugs in the refdb package that required custom scripts. These can be found below. This document is intended for downloading data for the COI gene widely used for animal DNA barcoding and metabarcoding.

#Installing and loading the refDB package (https://cran.r-project.org/web/packages/refdb/index.html)

The refdb package is available on CRAN, which means that you can easily install the latest stable version using the RStudio IDE or with the function install.packages. The other packages below will also be required to run refdb. If you are using a server to complete this step please check if the required softwore is already loaded. 

```{r}
#remotes::install_github("fkeck/refdb")
```

# load reqiured libraries
```{r message=FALSE, warning=FALSE}
suppressPackageStartupMessages({
library(refdb)
library(rentrez)
library(taxize)
library(bold)
library(tidyverse)
library(dplyr)
library(stringr)
library(here)})
```

# create folders for storing results
```{r}
# You can use the here::here function to define a path relative to the project 
#  directory or use an absolute path

raw_libraries_dir <- here::here("raw_libraries")
dir.create(raw_libraries_dir)

cleaned_libraries_dir <- here::here("cleaned_libraries")
dir.create(cleaned_libraries_dir)

final_library_dir <- here::here("final_library")
dir.create(final_library_dir)

geneious_library_dir <- here::here("final_library/geneious_library")
dir.create(geneious_library_dir)

trimming_dir <- here::here("final_library/trimming")
dir.create(trimming_dir)
```

--------------------------------------------------------------------------------
##CONTENTS

#SECTION 1*: Getting data from NCBI GenBank
Scripts for downloading data from the GenBank database and preparation of data for the reference database
  
#SECTION 2*: Getting data from BOLD
Scripts for downloading data from the BOLD database and preparation of data for the reference database
  
#SECTION 3: Preparing a private library as a DNA reference database
Private libraries are managed using sequencing editing software such as Geneious. Private libraries can be imported as .csv files and need to contain at a minimum : The lowest possible taxonomic classification (superkingdom;phylum;class;order;family;genus;species), an individual sequence identifier and a COI DNA sequence. Other metadata can be included but will not be used for reference library construction. As separate workflow document has been prepared for using BOLD to validate the identification in private libraries.

# SECTION 4: Creating a complete reference library 
Scripts for merging data from NCBI GenBank, BOLD and private libraries including final preparation and clean up. There are scripts available for exporting the complete final library as a .fasta flie which can be used aa a custom search database in Geneious.

#SECTION 5: Creating amplicon specific libraries for bioinformatic pipelines
These scripts produce a set of amplicon specific libraries that are in the correct format for use in bioinformatic pipelines. The scripts rely on the python package 'cutadapt' to trim the downloaded sequences to match amplicon sizes and positions. 
There are four amplicon libraries that can be produced for COI:
  - short amplicon (~220 bp using the BF1 /BR1 primer region from Elbrecht & Leese (2017))
  - right amplicon (~313 bp using the mICOIintF/dgHCO2198 primer region from Leray et al. (2013))
  - long amplicon (~422 bp using the BF2/BR2 primer region from Elbrecht & Leese (2017))
  - left amplicon (~316 bp using the B/R5 primer region from Hajibabaei et al. (2012))
Note: the short and right libraries are using for processing bulk samples (from the ARC linkage project)

* The download steps can take a few days to complete. If a download is not required (i.e., it was only completed in the last 6-12 months) previous data can be accessed as either the raw download or as a cleaned download (see below).


--------------------------------------------------------------------------------
# SECTION 1: Getting data from NCBI GenBank

The refdb packages uses the rentrez package (Winter 2017) to interface with NCBI servers. For animal DNA barcodes we will need to download data from the mitochondrial COI gene.

NOTE: This section is not currently working in refdb. These scripts are problematic for the download of data from NCBI GenBank (the link gets interupted). Jess has modified the code to make this step work but it is still problematic. Downloading NCBI COI sequences needs to be done over a few days. Don't expect everything to run smoothly. Expect errors and timeouts.

If a downloading from the NCBI GenBank data is not require please skip to 'Step 7' to import the last version of the cleaned NCBI GenBank data.

# set up
```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=TRUE, message=TRUE, error=TRUE, echo=TRUE, results="hold")
knitr::opts_knit$set(root.dir = "..")
options(digits=4)
options(width=120)
```

\
#Step 1: Set parameters/instructions for downloading data 

The refdb function `modified_refdb_import_NCBI` was split up into parts so downloading could be done in batches. Also some edits for taxon IDs that cause errors and keeping tmp files. Need to manually delete tmp files afterwards.

```{r}
modified_refdb_import_NCBI_pt1 <- function (query, max_seq_length = 10000, start = 0L, verbose = TRUE) 
{
    # ff <- tempfile("refdb_NCBI_", fileext = ".csv")
    # fx <- tempfile("refdb_NCBI_", fileext = ".xml")
    query <- paste0(query, " AND ( \"0\"[SLEN] : \"", max_seq_length, 
        "\"[SLEN] )")
    req <- rentrez::entrez_search(db = "nuccore", term = query, 
        use_history = TRUE)
    if (req$count == 0) {
        if (verbose) 
            cat("No sequence found\n")
        return(NULL)
    }
    if (verbose && start > 0L) 
        cat("Found", req$count, "sequences. Starting at", start, 
            ".\n")
    if (verbose) 
        # cat("Downloading", req$count - start, "sequences from NCBI...\n")
        cat("Found", req$count - start, "sequences from NCBI...\n")
    return(req)
}
    
modified_refdb_import_NCBI_pt2 <- function (request, csv_output, tmp_dir=tempdir(), full = FALSE, seq_bin = 200, 
    verbose = TRUE, start = 0L, end=NULL) 
{
    # Modified to write multiple csv files and keep them. Data frame is written twice for troubleshooting. Manually delete files after.
    if (! dir.exists(tmp_dir)) {
      dir.create(tmp_dir, recursive=TRUE)
    }
    if (is.null(end) || end > request$count) {
      end <- request$count
    }
    # ff_all <- tempfile("refdb_NCBI_", fileext = ".csv", tmpdir=tmp_dir)
    ff_all <- csv_output
    fx <- tempfile("refdb_NCBI_", fileext = ".xml", tmpdir=tmp_dir)
    for (seq_start in seq(start, end, seq_bin)) {
        ff <- tempfile(paste0("refdb_NCBI_", seq_start, "_"), fileext = ".csv", tmpdir=tmp_dir)
        recs <- refdb:::entrez_fetch_retry(db = "nuccore", web_history = request$web_history, 
            rettype = "gb", retmode = "xml", retmax = seq_bin, 
            retstart = seq_start, delay_retry = 60, n_retry = 50, 
            verbose = verbose)
        if (is.na(recs)) {
            next
        }
        readr::write_lines(recs, file = fx, append = FALSE)
        NCBI_xml <- xml2::read_xml(fx)
        NCBI_xml <- xml2::xml_children(NCBI_xml)
        NCBI_table <- refdb:::make_ncbi_table(NCBI_xml)
        taxo_id <- lapply(NCBI_xml, function(x) {
            res <- xml2::xml_find_all(x, ".//GBQualifier_name[text()=\"db_xref\"]/following-sibling::GBQualifier_value")
            res <- xml2::xml_text(res)
            res <- res[stringr::str_detect(res, "taxon:[0-9]+")]
            res <- unique(res)
            res <- stringr::str_extract(res, "(?<=taxon:)[0-9]+")
            # Sometimes a record doesn't have a taxon ID which causes an error downstream (e.g. MT491941)
  if (length(res) == 0) {
    message("Warning: Empty taxon id observed")
    res <- NA
  }
  return(res)
  })
taxo_id <- unlist(taxo_id)
taxo_id <- tibble::tibble(taxonomy = NCBI_table$taxonomy, 
                          id = taxo_id)

# Remove NA taxon IDs
taxo_id <- taxo_id[! is.na(taxo_id$id), ]

taxo_id <- taxo_id[!duplicated(taxo_id$taxonomy), ]
gtax <- get_ncbi_taxonomy_retry(taxo_id$id, delay_retry = 60, 
                                n_retry = 50, verbose = verbose)
taxo_id <- dplyr::left_join(taxo_id, gtax[, -ncol(gtax)], 
                            by = "id")
NCBI_table <- dplyr::left_join(NCBI_table, taxo_id, by = "taxonomy", 
                               suffix = c("", "_taxonomy"))
NCBI_table <- tibble::tibble(source = "NCBI", NCBI_table)
NCBI_table <- dplyr::mutate(NCBI_table, species = .data$organism)
if (full == FALSE) {
  NCBI_table <- dplyr::select(NCBI_table, .data$source, 
                              .data$id, .data$gene, .data$sequence, .data$superkingdom, 
                              .data$kingdom, .data$phylum, .data$subphylum, 
                              .data$class, .data$subclass, .data$infraclass, 
                              .data$order, .data$suborder, .data$infraorder, 
                              .data$superfamily, .data$family, .data$genus, 
                              .data$species, .data$country_location, .data$lat_lon)
}
if (seq_start == 0 | ! file.exists(ff_all)) {
  readr::write_csv(NCBI_table[0, ], file = ff_all)
}
readr::write_csv(NCBI_table, file = ff_all, append = TRUE, col_names = FALSE)
readr::write_csv(NCBI_table, file = ff, col_names = TRUE)
if (verbose) {
  cat("\r > ", seq_start + nrow(NCBI_table), " (", 
      round((seq_start + nrow(NCBI_table))/request$count * 
              100, digits = 1), "%) ", "sequences downloaded.", 
      sep = "")
}
}
}


modified_refdb_import_NCBI_pt3 <- function (csv_file) 
{
  ff <- csv_file
  out <- readr::read_csv(ff, col_types = readr::cols())
  out <- refdb:::process_geo_ncbi(out)
  out <- refdb_set_fields_NCBI(out)
  out <- refdb_set_fields(out, latitude = "latitude", longitude = "longitude")
  # file.remove(ff, fx)
  return(out)
}

# This is unedited from refdb:::get_ncbi_taxonomy_retry but needs to be here so I can edit get_ncbi_taxonomy
get_ncbi_taxonomy_retry <- function (id, delay_retry = 60, n_retry = 20, verbose = TRUE) 
{
  res <- "error"
  while (identical(res, "error") & n_retry > 0) {
    res <- tryCatch({
      Sys.sleep(0.1)
      get_ncbi_taxonomy(id, verbose = verbose)
    }, error = function(cond) {
      if (verbose) {
        message("\nSomething went wrong:")
        message(cond)
        message("\n")
        for (i in delay_retry:0) {
          cat("\rRetrying in", i, "s.  ")
          Sys.sleep(1)
        }
        cat("\n")
      }
      return("error")
    })
    n_retry <- n_retry - 1
  }
  if (identical(res, "error")) {
    stop("All attempts failed.")
  }
  else {
    return(res)
  }
}


xml_extract <- refdb:::xml_extract

get_ncbi_taxonomy <- function (id, verbose = TRUE) 
{
  ids <- split(id, ceiling(seq_along(id)/100))
  taxo_table <- lapply(ids, function(x) {
    taxo <- refdb:::entrez_fetch_retry("taxonomy", id = x, rettype = "xml", 
                                       verbose = verbose)
    taxo_xml <- xml2::read_xml(taxo)
    taxo_xml <- xml2::xml_children(taxo_xml)
    
    # This is super hacky...
    if (length(taxo_xml) != length(x)) {
      entrez_taxo_ids <- sapply(taxo_xml, function(y) {
        str_extract(xml2::xml_text(y), "^\\d+")
      })
      missing_ids <- x[! x %in% entrez_taxo_ids]
      message("Warning: taxon information from entrez has fewer records than the provided IDs. Removing ", missing_ids)
      x <- x[x %in% entrez_taxo_ids]
    }
    
    taxo_table <- tibble::tibble(id = x, superkingdom = xml_extract(taxo_xml, 
                                                                    ".//Rank[text()=\"superkingdom\"]/preceding-sibling::ScientificName"), 
                                 kingdom = xml_extract(taxo_xml, ".//Rank[text()=\"kingdom\"]/preceding-sibling::ScientificName"), 
                                 phylum = xml_extract(taxo_xml, ".//Rank[text()=\"phylum\"]/preceding-sibling::ScientificName"), 
                                 subphylum = xml_extract(taxo_xml, ".//Rank[text()=\"subphylum\"]/preceding-sibling::ScientificName"), 
                                 class = xml_extract(taxo_xml, ".//Rank[text()=\"class\"]/preceding-sibling::ScientificName"), 
                                 subclass = xml_extract(taxo_xml, ".//Rank[text()=\"subclass\"]/preceding-sibling::ScientificName"), 
                                 infraclass = xml_extract(taxo_xml, ".//Rank[text()=\"infraclass\"]/preceding-sibling::ScientificName"), 
                                 order = xml_extract(taxo_xml, ".//Rank[text()=\"order\"]/preceding-sibling::ScientificName"), 
                                 suborder = xml_extract(taxo_xml, ".//Rank[text()=\"suborder\"]/preceding-sibling::ScientificName"), 
                                 infraorder = xml_extract(taxo_xml, ".//Rank[text()=\"infraorder\"]/preceding-sibling::ScientificName"), 
                                 superfamily = xml_extract(taxo_xml, ".//Rank[text()=\"superfamily\"]/preceding-sibling::ScientificName"), 
                                 family = xml_extract(taxo_xml, ".//Rank[text()=\"family\"]/preceding-sibling::ScientificName"), 
                                 genus = xml_extract(taxo_xml, ".//Rank[text()=\"genus\"]/preceding-sibling::ScientificName"), 
                                 species = xml_extract(taxo_xml, ".//Rank[text()=\"species\"]/preceding-sibling::ScientificName"))
    return(taxo_table)
  })
  taxo_table <- dplyr::bind_rows(taxo_table)
  return(taxo_table)
}
```

# Step 2: Download sequence data for COI (the DNA barcode gene)
```{r}
Sys.time({
ncbi_request <- modified_refdb_import_NCBI_pt1("(((((((((COX1) OR cytochrome c oxidase I) OR COI) OR MT-CO1) OR MTCO1) OR CO I) OR mitochondrially encoded cytochrome c oxidase I) OR main subunit of cytochrome c oxidase) OR cytochrome c oxidase subunit I) AND Mitochondrion[Filter]")})
```

Note: The number of sequences change over time. e.g.  
4370766 sequences as of 2023-06-28.  
4371338 sequences as of 2023-06-29 and 30.  
4378270 sequences as of 2023-07-01.  

```{r}
seq(0, ncbi_request$count, by=3e5)
```

```{r eval=FALSE}
csv_file <- here::here("ncbi_2023-06-29.csv")
tmp_dir <- here::here("tmp_2023-06-29")

# e.g. Run commands in batches like so:
modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=0, end=300000)

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=300000, end=600000-1)

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=600000, end=900000-1)

# If things break, can manually specify start and end indices. Duplicate sequences will be removed at a later step, so don't worry about downloading sequences twice.
```

```{r eval=FALSE}
csv_file <- here("ncbi_2023-06-30.csv")
tmp_dir <- here("tmp_2023-06-30")

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=900000, end=1200000-1)

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=1200000, end=1500000-1)
```

```{r eval=FALSE}
csv_file <- here("ncbi_2023-07-01.csv")
tmp_dir <- here("tmp_2023-07-01")

# If the database updates overnight, you can offset the start index by some amount (e.g.)
modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=1500000-7000, end=1800000-1)

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=1800000, end=2100000-1)

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=2100000, end=2400000-1)

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=2400000, end=3000000-1)

```

```{r eval=FALSE}
csv_file <- here("ncbi_2023-07-02.csv")
tmp_dir <- here("tmp_2023-07-02")

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=3000000, end=3300000-1)

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=3300000, end=3900000-1)

modified_refdb_import_NCBI_pt2(ncbi_request, csv_output = csv_file, tmp_dir = tmp_dir, start=3900000)
```

# Load in files
```{r}
# Overlap between june 28 and 29, and june 30 and july 01
f1 <- modified_refdb_import_NCBI_pt3(here("ncbi_2023-06-28.csv"))
f2 <- modified_refdb_import_NCBI_pt3(here("ncbi_2023-06-29.csv"))
f3 <- modified_refdb_import_NCBI_pt3(here("ncbi_2023-06-30.csv"))
f4 <- modified_refdb_import_NCBI_pt3(here("ncbi_2023-07-01.csv"))
f5 <- modified_refdb_import_NCBI_pt3(here("ncbi_2023-07-02.csv"))
```

```{r}
# Combine everything
combined_with_dups <- rbind(f1, f2, f3, f4, f5)
nrow(combined_with_dups)
table(duplicated(combined_with_dups$id))

# Remove duplicates
dups <- rev(duplicated(rev(combined_with_dups$id)))
combined <- combined_with_dups[! dups,]
table(duplicated(combined$id))

# I think we should name the file with the database name and date (i.e., all NCBI_Oct23)
saveRDS(combined, file=here::here("ref_db/combined_tbl_df.rds"))
rm(combined_with_dups)
```

# Session Info
  
```{r}
if (nzchar(system.file(package="devtools"))) {
  devtools::session_info()
} else {
  sessionInfo()
}
```

^ The section above for downloading sequences from NCBI GenBank has been configured by Jess. Mel has not run through this yet.

# Step 3: Retriving the last saved version of NCBI GenBank raw COI data file (if starting from this point)
```{r}
ncbi_data <- readRDS("/mnt/galaxy/home/jess/projects/metabarcoding_pipeline/ref_db/combined_tbl_df.rds")
```

# Step 4: Preparing NCBI GenBank file
Some records may have no sequence data, so need to be removed. Also, BOLD data on NCBI is named 'sp. BOLD:' should be converted to 'sp. B-' to be consistent with the naming in the mwbug database and our private DNA barcode library (from our own DNA barcoding). This data needs to be split from other records and the 'extra bits on names' filter used on the dataset without BOLD id's. The fields need to be set for refdb.

```{r}
#remove files with no sequence data
ncbi_data <- ncbi_data[!is.na(ncbi_data$sequence),]

tax_blank <- ncbi_data[is.na(ncbi_data$superkingdom),]

tax_blank <- ncbi_data[!is.na(ncbi_data$superkingdom),]
tax_blank <- ncbi_data[grep("NA", ncbi_data$species),] 

# group records with BOLD BIN's
ncbi_bold <- ncbi_data[grep("sp. BOLD:", ncbi_data$species),] 
# adjust names to be inline with mw_bugs
ncbi_bold$species <- gsub("sp. BOLD:", "sp. B-", ncbi_bold$species)

# group records without BOLD BIN's
ncbi_nonbold <- ncbi_data[-grep("sp. BOLD:", ncbi_data$species),]
# remove extra bits on names which are added by NCBI
ncbi_nonbold <- refdb::refdb_clean_tax_remove_extra(ncbi_nonbold)

#merge back together
ncbi_data_clean <- refdb::refdb_merge(ncbi_bold, ncbi_nonbold)

# set fields for refdb
ncbi_data_clean <- refdb::refdb_set_fields(ncbi_data_clean,
                              taxonomy = c(kingdom = "superkingdom",
                                phylum = "phylum",
                                class = "class",
                                order = "order",
                                family = "family",
                                genus = "genus",
                                species = "species"),
                              sequence = "sequence",
                              marker = "gene",
                              source = "source",
                              id = "id")

refdb::refdb_get_fields(ncbi_data_clean)

#remove files no longer needed
rm(ncbi_bold)
rm(ncbi_nonbold)
rm(ncbi_data)
```

## Cleaning DNA sequences and taxonomic data

# Step 4: Remove duplicates
This will remove records with identical sequence and taxonomic information. Expect the number of sequence records to reduce by 30-50%.
```{r}
ncbi_data_clean <- refdb::refdb_filter_seq_duplicates(ncbi_data_clean)
```

# Step 6: Plot sequence length of NCBI data
This will give an idea of the sequence length of the data retrieved from NCBI. The full COI gene is ~1500 bp long in animals. DNA barcodes are ~658bp but can vary.

```{r}
refdb::refdb_plot_seqlen_hist(NCBI_data_clean)
```

# Visualizing the reference library (Optional)
Now letâ€™s take a tour of the functions you can use to produce graphical representation of your reference database. Because refdb stores reference database as dataframes it is straightforward to produce plots (e.g. with the tidyverse and ggplot2). For example, we can make a barplot showing the distribution of the phyla like this:
  
```{r}
ncbi_data_clean %>% 
  group_by(phylum) %>% 
  count() %>% 
  ggplot(aes(fct_reorder(phylum, n, .desc = TRUE), n)) +
  geom_col() +
  xlab("Phylum") +
  ylab("Number of records") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0.5))
```

Additionally, refdb provides ready-to-use functions to produce more sophisticated plots. For example, to represent the taxonomic coverage of your reference library across multiple levels with a tree map:
  
```{r}
refdb::refdb_plot_tax_treemap(ncbi_data_clean, cols = 2)
```

Alternatively you can represent this information with a taxonomic tree. These functions have several parameters to control what is represented in the plot (taxonomic ranks, colors, etc.).

# Creating a report (useful)
A report is a simple and rapid way to get an overview of the current state of your reference library and to identify some possible issues. You can compile a report using the function refdb_report: 

```{r}
refdb::refdb_report(ncbi_data_clean, view=TRUE)
```                       

The result (not shown here) is an interactive HTML report that can be opened in any recent web browser. It contains some statistics and plots, and the results of functions refdb_check_tax_typo, refdb_check_tax_conflict and refdb_check_seq_conflict. These functions can be used to identify possible spelling errors in taxonomic names, conflicts in the taxonomic tree and lack of genetic resolution in sequences, respectively.

# Step 7: Export cleaned data (add source database, status and date to file name)
This data will be used to create the DNA barcode reference library and will be combined with the data  below which includes downloads from BOLD and data from our own DNA sequencing (private DNA barcodes)

```{r}
# save cleaned data
saveRDS(ncbi_data_clean, file=here::here("/cleaned_libraries/NCBI_data_clean_Oct23.rds"))

# importing cleaned NCBI GenBank data (if required)

#load cleaned data
#ncbi_data_clean <- readRDS(here::here("cleaned_libraries/NCBI_data_clean_Oct23.rds"))
```

--------------------------------------------------------------------------------
# SECTION 2: Getting data from BOLD

You may have notice that the search interface is a bit different. This is because here we rely on another package (bold, Chamberlain 2020) to download the data. You can check the manual of refdb_import_BOLD to see the different arguments available.

Data needs downloaded for animals groups at different taxonomic levels depending on the number of sequences in that group. For small sequence groups, it can be downloaded at phlyum level for larger sequence groups with more sequence records, we'll need to follow the BOLD "large data" instructions which limits the size of data download requests (see ?refdb_import_BOLD). I have configured this below using trial and error to enable data from all animal groups to be download. I have also included scripts for proteobactera, plant and fungi so these group can be identified and excluded.

NOTE: if a download of the data is not require please skip to 'Step 9' import the last version of BOLD data downloaded and cleaned.

# Step 1: Download data from the BOLD database
```{r}
# Invertebrate that can be downloaded at phylum level
Annelida_bold <- refdb::refdb_import_BOLD(taxon = "Annelida", ncbi_taxo = FALSE)
Cnidaria_bold <- refdb::refdb_import_BOLD(taxon = "Cnidaria", ncbi_taxo = FALSE)
Bryozoa_bold <- refdb::refdb_import_BOLD(taxon = "Bryozoa", ncbi_taxo = FALSE)
Platyhelminthes_bold <- refdb::refdb_import_BOLD(taxon = "Platyhelminthes", ncbi_taxo = FALSE)
Porifera_bold <- refdb::refdb_import_BOLD(taxon = "Porifera", ncbi_taxo = FALSE)
Nematoda_bold <- refdb::refdb_import_BOLD(taxon = "Nematoda", ncbi_taxo = FALSE)
Rotifera_bold <- refdb::refdb_import_BOLD(taxon = "Rotifera", ncbi_taxo = FALSE)
Nemertea_bold <- refdb::refdb_import_BOLD(taxon = "Nemertea", ncbi_taxo = FALSE)
Onychophora_bold <- refdb::refdb_import_BOLD(taxon = "Onychophora", ncbi_taxo = FALSE)
Priapulida_bold <- refdb::refdb_import_BOLD(taxon = "Priapulida", ncbi_taxo = FALSE)
Tardigrada_bold <- refdb::refdb_import_BOLD(taxon = "Tardigrada", ncbi_taxo = FALSE)
Mollusca_bold <- refdb::refdb_import_BOLD(taxon = "Mollusca", ncbi_taxo = FALSE)

#All remaining groups are considered 'large datasets' and will not download from BOLD unless they are broken down into smaller data sets. This can be done by using low taxonomic levels i.e. class, order or family.

# Invertebrates that can be downloaded at class level - I have left out marine only classes and Insecta (which is too large - see below)
Merostomata_bold <- refdb::refdb_import_BOLD(taxon = "Merostomata", ncbi_taxo = FALSE)
Pauropoda_bold <- refdb::refdb_import_BOLD(taxon = "Pauropoda", ncbi_taxo = FALSE)
Symphyla_bold <- refdb::refdb_import_BOLD(taxon = "Symphyla", ncbi_taxo = FALSE)
Diplopoda_bold <- refdb::refdb_import_BOLD(taxon = "Diplopoda", ncbi_taxo = FALSE)
Chilopoda_bold <- refdb::refdb_import_BOLD(taxon = "Chilopoda", ncbi_taxo = FALSE)
Copepoda_bold <- refdb::refdb_import_BOLD(taxon = "Copepoda", ncbi_taxo = FALSE)
Collembola_bold <- refdb::refdb_import_BOLD(taxon = "Collembola", ncbi_taxo = FALSE)
Protura_bold <- refdb::refdb_import_BOLD(taxon = "Protura", ncbi_taxo = FALSE)
Cephalocarida_bold <- refdb::refdb_import_BOLD(taxon = "Cephalocarida", ncbi_taxo = FALSE)
Remipedia_bold <- refdb::refdb_import_BOLD(taxon = "Remipedia", ncbi_taxo = FALSE)
Branchiopoda_bold <- refdb::refdb_import_BOLD(taxon = "Branchiopoda", ncbi_taxo = FALSE)
Thecostraca_bold <- refdb::refdb_import_BOLD(taxon = "Thecostraca", ncbi_taxo = FALSE)
Malacostraca_bold <- refdb::refdb_import_BOLD(taxon = "Malacostraca", ncbi_taxo = FALSE)
Ichthyostraca_bold <- refdb::refdb_import_BOLD(taxon = "Ichthyostraca", ncbi_taxo = FALSE)
Diplura_bold <- refdb::refdb_import_BOLD(taxon = "Diplura", ncbi_taxo = FALSE)
Pycnogonida_bold <- refdb::refdb_import_BOLD(taxon = "Pycnogonida", ncbi_taxo = FALSE)
Ostracoda_bold <- refdb::refdb_import_BOLD(taxon = "Ostracoda", ncbi_taxo = FALSE)
Arachnida_bold <- refdb::refdb_import_BOLD(taxon = "Arachnida", ncbi_taxo = FALSE)

# Invertebrate that can be downloaded at order level - includes Insecta (except Diptera (which is too large -see below))
Archaeognatha_bold <- refdb::refdb_import_BOLD(taxon = "Archaeognatha", ncbi_taxo = FALSE)
Blattodea_bold <- refdb::refdb_import_BOLD(taxon = "Blattodea", ncbi_taxo = FALSE)
Coleoptera_bold <- refdb::refdb_import_BOLD(taxon = "Coleoptera", ncbi_taxo = FALSE)
Dermaptera_bold <- refdb::refdb_import_BOLD(taxon = "Dermaptera", ncbi_taxo = FALSE)
Embioptera_bold <- refdb::refdb_import_BOLD(taxon = " Embioptera", ncbi_taxo = FALSE)
Ephemeroptera_bold <- refdb::refdb_import_BOLD(taxon = " Ephemeroptera", ncbi_taxo = FALSE)
Hemiptera_bold <- refdb::refdb_import_BOLD(taxon = "Hemiptera", ncbi_taxo = FALSE)
Hymenoptera_bold <- refdb::refdb_import_BOLD(taxon = "Hymenoptera", ncbi_taxo = FALSE)
Lepidoptera_bold <- refdb::refdb_import_BOLD(taxon = "Lepidoptera", ncbi_taxo = FALSE)
Mantodea_bold <- refdb::refdb_import_BOLD(taxon = "Mantodea", ncbi_taxo = FALSE)
Mecoptera_bold <- refdb::refdb_import_BOLD(taxon = "Mecoptera", ncbi_taxo = FALSE)
Megaloptera_bold <- refdb::refdb_import_BOLD(taxon = "Megaloptera", ncbi_taxo = FALSE)
Neuroptera_bold <- refdb::refdb_import_BOLD(taxon = "Neuroptera", ncbi_taxo = FALSE)
Notoptera_bold <- refdb::refdb_import_BOLD(taxon = "Notoptera", ncbi_taxo = FALSE)
Odonata_bold <- refdb::refdb_import_BOLD(taxon = "Odonata", ncbi_taxo = FALSE)
Orthoptera_bold <- refdb::refdb_import_BOLD(taxon = "Orthoptera", ncbi_taxo = FALSE)
Phasmatodea_bold <- refdb::refdb_import_BOLD(taxon = "Phasmatodea", ncbi_taxo = FALSE)
Plecoptera_bold <- refdb::refdb_import_BOLD(taxon = "Plecoptera", ncbi_taxo = FALSE)
Psocodea_bold <- refdb::refdb_import_BOLD(taxon = "Psocodea", ncbi_taxo = FALSE)
Raphidioptera_bold <- refdb::refdb_import_BOLD(taxon = "Raphidioptera", ncbi_taxo = FALSE)
Siphonaptera_bold <- refdb::refdb_import_BOLD(taxon = "Siphonaptera", ncbi_taxo = FALSE)
Strepsiptera_bold <- refdb::refdb_import_BOLD(taxon = "Strepsiptera", ncbi_taxo = FALSE)
Thysanoptera_bold <- refdb::refdb_import_BOLD(taxon = "Thysanoptera", ncbi_taxo = FALSE)
Trichoptera_bold <- refdb::refdb_import_BOLD(taxon = "Trichoptera", ncbi_taxo = FALSE)
Zoraptera_bold <- refdb::refdb_import_BOLD(taxon = "Zoraptera", ncbi_taxo = FALSE)
Zygentoma_bold <- refdb::refdb_import_BOLD(taxon = "Zygentoma", ncbi_taxo = FALSE)

# Invertebrates that can be downloaded at family - for Diptera families
Acartophthalmidae_bold <- refdb::refdb_import_BOLD(taxon = "Acartophthalmidae" , ncbi_taxo = FALSE)
Acroceridae_bold <- refdb::refdb_import_BOLD(taxon = "Acroceridae" , ncbi_taxo = FALSE)
Agromyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Agromyzidae" , ncbi_taxo = FALSE)
Anisopodidae_bold <- refdb::refdb_import_BOLD(taxon = "Anisopodidae" , ncbi_taxo = FALSE)
Anthomyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Anthomyiidae" , ncbi_taxo = FALSE)
Anthomyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Anthomyzidae" , ncbi_taxo = FALSE)
Apioceridae_bold <- refdb::refdb_import_BOLD(taxon = "Apioceridae" , ncbi_taxo = FALSE)
Apsilocephalidae_bold <- refdb::refdb_import_BOLD(taxon = "Apsilocephalidae" , ncbi_taxo = FALSE)
Apystomyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Apystomyiidae" , ncbi_taxo = FALSE)
Asilidae_bold <- refdb::refdb_import_BOLD(taxon = "Asilidae" , ncbi_taxo = FALSE)
Asteiidae_bold <- refdb::refdb_import_BOLD(taxon = "Asteiidae" , ncbi_taxo = FALSE)
Atelestidae_bold <- refdb::refdb_import_BOLD(taxon = "Atelestidae" , ncbi_taxo = FALSE)
Athericidae_bold <- refdb::refdb_import_BOLD(taxon = "Athericidae" , ncbi_taxo = FALSE)
Aulacigastridae_bold <- refdb::refdb_import_BOLD(taxon = "Aulacigastridae" , ncbi_taxo = FALSE)
Australimyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Australimyzidae" , ncbi_taxo = FALSE)
Austroleptidae_bold <- refdb::refdb_import_BOLD(taxon = "Austroleptidae" , ncbi_taxo = FALSE)
Bibionidae_bold <- refdb::refdb_import_BOLD(taxon = "Bibionidae" , ncbi_taxo = FALSE)
Blephariceridae_bold <- refdb::refdb_import_BOLD(taxon = "Blephariceridae" , ncbi_taxo = FALSE)
Bolbomyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Bolbomyiidae" , ncbi_taxo = FALSE)
Bolitophilidae_bold <- refdb::refdb_import_BOLD(taxon = "Bolitophilidae" , ncbi_taxo = FALSE)
Bombyliidae_bold <- refdb::refdb_import_BOLD(taxon = "Bombyliidae" , ncbi_taxo = FALSE)
Braulidae_bold <- refdb::refdb_import_BOLD(taxon = "Braulidae" , ncbi_taxo = FALSE)
Calliphoridae_bold <- refdb::refdb_import_BOLD(taxon = "Calliphoridae" , ncbi_taxo = FALSE)
Camillidae_bold <- refdb::refdb_import_BOLD(taxon = "Camillidae" , ncbi_taxo = FALSE)
Canacidae_bold <- refdb::refdb_import_BOLD(taxon = "Canacidae" , ncbi_taxo = FALSE)
Canthyloscelidae_bold <- refdb::refdb_import_BOLD(taxon = "Canthyloscelidae" , ncbi_taxo = FALSE)
Carnidae_bold <- refdb::refdb_import_BOLD(taxon = "Carnidae" , ncbi_taxo = FALSE)
Cecidomyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Cecidomyiidae" , ncbi_taxo = FALSE)
Celyphidae_bold <- refdb::refdb_import_BOLD(taxon = "Celyphidae" , ncbi_taxo = FALSE)
Ceratopogonidae_bold <- refdb::refdb_import_BOLD(taxon = "Ceratopogonidae" , ncbi_taxo = FALSE)
Chamaemyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Chamaemyiidae" , ncbi_taxo = FALSE)
Chaoboridae_bold <- refdb::refdb_import_BOLD(taxon = "Chaoboridae" , ncbi_taxo = FALSE)
Chironomidae_bold <- refdb::refdb_import_BOLD(taxon = "Chironomidae" , ncbi_taxo = FALSE)
Chloropidae_bold <- refdb::refdb_import_BOLD(taxon = "Chloropidae" , ncbi_taxo = FALSE)
Chyromyidae_bold <- refdb::refdb_import_BOLD(taxon = "Chyromyidae" , ncbi_taxo = FALSE)
Clusiidae_bold <- refdb::refdb_import_BOLD(taxon = "Clusiidae" , ncbi_taxo = FALSE)
Coelopidae_bold <- refdb::refdb_import_BOLD(taxon = "Coelopidae" , ncbi_taxo = FALSE)
Conopidae_bold <- refdb::refdb_import_BOLD(taxon = "Conopidae" , ncbi_taxo = FALSE)
Corethrellidae_bold <- refdb::refdb_import_BOLD(taxon = "Corethrellidae" , ncbi_taxo = FALSE)
Cryptochetidae_bold <- refdb::refdb_import_BOLD(taxon = "Cryptochetidae" , ncbi_taxo = FALSE)
Ctenostylidae_bold <- refdb::refdb_import_BOLD(taxon = "Ctenostylidae" , ncbi_taxo = FALSE)
Culicidae_bold <- refdb::refdb_import_BOLD(taxon = "Culicidae" , ncbi_taxo = FALSE)
Curtonotidae_bold <- refdb::refdb_import_BOLD(taxon = "Curtonotidae" , ncbi_taxo = FALSE)
Cylindrotomidae_bold <- refdb::refdb_import_BOLD(taxon = "Cylindrotomidae" , ncbi_taxo = FALSE)
Cypselosomatidae_bold <- refdb::refdb_import_BOLD(taxon = "Cypselosomatidae" , ncbi_taxo = FALSE)
Deuterophlebiidae_bold <- refdb::refdb_import_BOLD(taxon = "Deuterophlebiidae" , ncbi_taxo = FALSE)
Diadocidiidae_bold <- refdb::refdb_import_BOLD(taxon = "Diadocidiidae" , ncbi_taxo = FALSE)
Diastatidae_bold <- refdb::refdb_import_BOLD(taxon = "Diastatidae" , ncbi_taxo = FALSE)
Diopsidae_bold <- refdb::refdb_import_BOLD(taxon = "Diopsidae" , ncbi_taxo = FALSE)
Diptera_family_incertae_sedis_bold <- refdb::refdb_import_BOLD(taxon = "Diptera_family_incertae_sedis" , ncbi_taxo = FALSE)
Ditomyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Ditomyiidae" , ncbi_taxo = FALSE)
Dixidae_bold <- refdb::refdb_import_BOLD(taxon = "Dixidae" , ncbi_taxo = FALSE)
Dolichopodidae_bold <- refdb::refdb_import_BOLD(taxon = "Dolichopodidae" , ncbi_taxo = FALSE)
Drosophilidae_bold <- refdb::refdb_import_BOLD(taxon = "Drosophilidae" , ncbi_taxo = FALSE)
Dryomyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Dryomyzidae" , ncbi_taxo = FALSE)
Empididae_bold <- refdb::refdb_import_BOLD(taxon = "Empididae" , ncbi_taxo = FALSE)
Ephydridae_bold <- refdb::refdb_import_BOLD(taxon = "Ephydridae" , ncbi_taxo = FALSE)
Evocoidae_bold <- refdb::refdb_import_BOLD(taxon = "Evocoidae" , ncbi_taxo = FALSE)
Fanniidae_bold <- refdb::refdb_import_BOLD(taxon = "Fanniidae" , ncbi_taxo = FALSE)
Fergusoninidae_bold <- refdb::refdb_import_BOLD(taxon = "Fergusoninidae" , ncbi_taxo = FALSE)
Glossinidae_bold <- refdb::refdb_import_BOLD(taxon = "Glossinidae" , ncbi_taxo = FALSE)
Gobryidae_bold <- refdb::refdb_import_BOLD(taxon = "Gobryidae" , ncbi_taxo = FALSE)
Helcomyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Helcomyzidae" , ncbi_taxo = FALSE)
Heleomyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Heleomyzidae" , ncbi_taxo = FALSE)
Helosciomyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Helosciomyzidae" , ncbi_taxo = FALSE)
Hesperinidae_bold <- refdb::refdb_import_BOLD(taxon = "Hesperinidae" , ncbi_taxo = FALSE)
Hilarimorphidae_bold <- refdb::refdb_import_BOLD(taxon = "Hilarimorphidae" , ncbi_taxo = FALSE)
Hippoboscidae_bold <- refdb::refdb_import_BOLD(taxon = "Hippoboscidae" , ncbi_taxo = FALSE)
Homalocnemidae_bold <- refdb::refdb_import_BOLD(taxon = "Homalocnemidae" , ncbi_taxo = FALSE)
Huttoninidae_bold <- refdb::refdb_import_BOLD(taxon = "Huttoninidae" , ncbi_taxo = FALSE)
Hybotidae_bold <- refdb::refdb_import_BOLD(taxon = "Hybotidae" , ncbi_taxo = FALSE)
Inbiomyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Inbiomyiidae" , ncbi_taxo = FALSE)
Ironomyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Ironomyiidae" , ncbi_taxo = FALSE)
Iteaphilidae_bold <- refdb::refdb_import_BOLD(taxon = "Iteaphilidae" , ncbi_taxo = FALSE)
Keroplatidae_bold <- refdb::refdb_import_BOLD(taxon = "Keroplatidae" , ncbi_taxo = FALSE)
Lauxaniidae_bold <- refdb::refdb_import_BOLD(taxon = "Lauxaniidae" , ncbi_taxo = FALSE)
Limoniidae_bold <- refdb::refdb_import_BOLD(taxon = "Limoniidae" , ncbi_taxo = FALSE)
Lonchaeidae_bold <- refdb::refdb_import_BOLD(taxon = "Lonchaeidae" , ncbi_taxo = FALSE)
Lonchopteridae_bold <- refdb::refdb_import_BOLD(taxon = "Lonchopteridae" , ncbi_taxo = FALSE)
Marginidae_bold <- refdb::refdb_import_BOLD(taxon = "Marginidae" , ncbi_taxo = FALSE)
Megamerinidae_bold <- refdb::refdb_import_BOLD(taxon = "Megamerinidae" , ncbi_taxo = FALSE)
Mesembrinellidae_bold <- refdb::refdb_import_BOLD(taxon = "Mesembrinellidae" , ncbi_taxo = FALSE)
Micropezidae_bold <- refdb::refdb_import_BOLD(taxon = "Micropezidae" , ncbi_taxo = FALSE)
Milichiidae_bold <- refdb::refdb_import_BOLD(taxon = "Milichiidae" , ncbi_taxo = FALSE)
Mormotomyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Mormotomyiidae" , ncbi_taxo = FALSE)
Muscidae_bold <- refdb::refdb_import_BOLD(taxon = "Muscidae" , ncbi_taxo = FALSE)
Mycetophilidae_bold <- refdb::refdb_import_BOLD(taxon = "Mycetophilidae" , ncbi_taxo = FALSE)
Mydidae_bold <- refdb::refdb_import_BOLD(taxon = "Mydidae" , ncbi_taxo = FALSE)
Mystacinobiidae_bold <- refdb::refdb_import_BOLD(taxon = "Mystacinobiidae" , ncbi_taxo = FALSE)
Nannodastiidae_bold <- refdb::refdb_import_BOLD(taxon = "Nannodastiidae" , ncbi_taxo = FALSE)
Natalimyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Natalimyzidae" , ncbi_taxo = FALSE)
Nemestrinidae_bold <- refdb::refdb_import_BOLD(taxon = "Nemestrinidae" , ncbi_taxo = FALSE)
Neminidae_bold <- refdb::refdb_import_BOLD(taxon = "Neminidae" , ncbi_taxo = FALSE)
Neriidae_bold <- refdb::refdb_import_BOLD(taxon = "Neriidae" , ncbi_taxo = FALSE)
Neurochaetidae_bold <- refdb::refdb_import_BOLD(taxon = "Neurochaetidae" , ncbi_taxo = FALSE)
Nothybidae_bold <- refdb::refdb_import_BOLD(taxon = "Nothybidae" , ncbi_taxo = FALSE)
Nymphomyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Nymphomyiidae" , ncbi_taxo = FALSE)
Odiniidae_bold <- refdb::refdb_import_BOLD(taxon = "Odiniidae" , ncbi_taxo = FALSE)
Oestridae_bold <- refdb::refdb_import_BOLD(taxon = "Oestridae" , ncbi_taxo = FALSE)
Opetiidae_bold <- refdb::refdb_import_BOLD(taxon = "Opetiidae" , ncbi_taxo = FALSE)
Opomyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Opomyzidae" , ncbi_taxo = FALSE)
Oreogetonidae_bold <- refdb::refdb_import_BOLD(taxon = "Oreogetonidae" , ncbi_taxo = FALSE)
Oreoleptidae_bold <- refdb::refdb_import_BOLD(taxon = "Oreoleptidae" , ncbi_taxo = FALSE)
Pachyneuridae_bold <- refdb::refdb_import_BOLD(taxon = "Pachyneuridae" , ncbi_taxo = FALSE)
Pallopteridae_bold <- refdb::refdb_import_BOLD(taxon = "Pallopteridae" , ncbi_taxo = FALSE)
Pantophthalmidae_bold <- refdb::refdb_import_BOLD(taxon = "Pantophthalmidae" , ncbi_taxo = FALSE)
Paraleucopidae_bold <- refdb::refdb_import_BOLD(taxon = "Paraleucopidae" , ncbi_taxo = FALSE)
Pediciidae_bold <- refdb::refdb_import_BOLD(taxon = "Pediciidae" , ncbi_taxo = FALSE)
Pelecorhynchidae_bold <- refdb::refdb_import_BOLD(taxon = "Pelecorhynchidae" , ncbi_taxo = FALSE)
Periscelididae_bold <- refdb::refdb_import_BOLD(taxon = "Periscelididae" , ncbi_taxo = FALSE)
Perissommatidae_bold <- refdb::refdb_import_BOLD(taxon = "Perissommatidae" , ncbi_taxo = FALSE)
Phoridae_bold <- refdb::refdb_import_BOLD(taxon = "Phoridae" , ncbi_taxo = FALSE)
Piophilidae_bold <- refdb::refdb_import_BOLD(taxon = "Piophilidae" , ncbi_taxo = FALSE)
Pipunculidae_bold <- refdb::refdb_import_BOLD(taxon = "Pipunculidae" , ncbi_taxo = FALSE)
Platypezidae_bold <- refdb::refdb_import_BOLD(taxon = "Platypezidae" , ncbi_taxo = FALSE)
Platystomatidae_bold <- refdb::refdb_import_BOLD(taxon = "Platystomatidae" , ncbi_taxo = FALSE)
Polleniidae_bold <- refdb::refdb_import_BOLD(taxon = "Polleniidae" , ncbi_taxo = FALSE)
Pseudopomyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Pseudopomyzidae" , ncbi_taxo = FALSE)
Psilidae_bold <- refdb::refdb_import_BOLD(taxon = "Psilidae" , ncbi_taxo = FALSE)
Psychodidae_bold <- refdb::refdb_import_BOLD(taxon = "Psychodidae" , ncbi_taxo = FALSE)
Ptychopteridae_bold <- refdb::refdb_import_BOLD(taxon = "Ptychopteridae" , ncbi_taxo = FALSE)
Pyrgotidae_bold <- refdb::refdb_import_BOLD(taxon = "Pyrgotidae" , ncbi_taxo = FALSE)
Rhagionidae_bold <- refdb::refdb_import_BOLD(taxon = "Rhagionidae" , ncbi_taxo = FALSE)
Richardiidae_bold <- refdb::refdb_import_BOLD(taxon = "Richardiidae" , ncbi_taxo = FALSE)
Ropalomeridae_bold <- refdb::refdb_import_BOLD(taxon = "Ropalomeridae" , ncbi_taxo = FALSE)
Sarcophagidae_bold <- refdb::refdb_import_BOLD(taxon = "Sarcophagidae" , ncbi_taxo = FALSE)
Scathophagidae_bold <- refdb::refdb_import_BOLD(taxon = "Scathophagidae" , ncbi_taxo = FALSE)
Scatopsidae_bold <- refdb::refdb_import_BOLD(taxon = "Scatopsidae" , ncbi_taxo = FALSE)
Scenopinidae_bold <- refdb::refdb_import_BOLD(taxon = "Scenopinidae" , ncbi_taxo = FALSE)
Sciaridae_bold <- refdb::refdb_import_BOLD(taxon = "Sciaridae" , ncbi_taxo = FALSE)
Sciaroidea_incertae_sedis_bold <- refdb::refdb_import_BOLD(taxon = "Sciaroidea_incertae_sedis" , ncbi_taxo = FALSE)
Sciomyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Sciomyzidae" , ncbi_taxo = FALSE)
Sepsidae_bold <- refdb::refdb_import_BOLD(taxon = "Sepsidae" , ncbi_taxo = FALSE)
Simuliidae_bold <- refdb::refdb_import_BOLD(taxon = "Simuliidae" , ncbi_taxo = FALSE)
Somatiidae_bold <- refdb::refdb_import_BOLD(taxon = "Somatiidae" , ncbi_taxo = FALSE)
Sphaeroceridae_bold <- refdb::refdb_import_BOLD(taxon = "Sphaeroceridae" , ncbi_taxo = FALSE)
Stratiomyidae_bold <- refdb::refdb_import_BOLD(taxon = "Stratiomyidae" , ncbi_taxo = FALSE)
Strongylophthalmyiidae_bold <- refdb::refdb_import_BOLD(taxon = "Strongylophthalmyiidae" , ncbi_taxo = FALSE)
Syringogastridae_bold <- refdb::refdb_import_BOLD(taxon = "Syringogastridae" , ncbi_taxo = FALSE)
Syrphidae_bold <- refdb::refdb_import_BOLD(taxon = "Syrphidae" , ncbi_taxo = FALSE)
Tabanidae_bold <- refdb::refdb_import_BOLD(taxon = "Tabanidae" , ncbi_taxo = FALSE)
Tachinidae_bold <- refdb::refdb_import_BOLD(taxon = "Tachinidae" , ncbi_taxo = FALSE)
Tanyderidae_bold <- refdb::refdb_import_BOLD(taxon = "Tanyderidae" , ncbi_taxo = FALSE)
Tanypezidae_bold <- refdb::refdb_import_BOLD(taxon = "Tanypezidae" , ncbi_taxo = FALSE)
Tephritidae_bold <- refdb::refdb_import_BOLD(taxon = "Tephritidae" , ncbi_taxo = FALSE)
Teratomyzidae_bold <- refdb::refdb_import_BOLD(taxon = "Teratomyzidae" , ncbi_taxo = FALSE)
Thaumaleidae_bold <- refdb::refdb_import_BOLD(taxon = "Thaumaleidae" , ncbi_taxo = FALSE)
Therevidae_bold <- refdb::refdb_import_BOLD(taxon = "Therevidae" , ncbi_taxo = FALSE)
Tipulidae_bold <- refdb::refdb_import_BOLD(taxon = "Tipulidae" , ncbi_taxo = FALSE)
Trichoceridae_bold <- refdb::refdb_import_BOLD(taxon = "Trichoceridae" , ncbi_taxo = FALSE)
Ulidiidae_bold <- refdb::refdb_import_BOLD(taxon = "Ulidiidae" , ncbi_taxo = FALSE)
Vermileonidae_bold <- refdb::refdb_import_BOLD(taxon = "Vermileonidae" , ncbi_taxo = FALSE)
Xenasteiidae_bold <- refdb::refdb_import_BOLD(taxon = "Xenasteiidae" , ncbi_taxo = FALSE)
Xylomyidae_bold <- refdb::refdb_import_BOLD(taxon = "Xylomyidae" , ncbi_taxo = FALSE)
Xylophagidae_bold <- refdb::refdb_import_BOLD(taxon = "Xylophagidae" , ncbi_taxo = FALSE) 

# Download for vertebrates barcodes (these may also turn up in our samples via eDNA or by-catch)
Amphibia_bold <- refdb::refdb_import_BOLD(taxon = "	Amphibia", ncbi_taxo = FALSE)
Leptocardii_bold <- refdb::refdb_import_BOLD(taxon = "Leptocardii", ncbi_taxo = FALSE)
Mammalia_bold <- refdb::refdb_import_BOLD(taxon = "Mammalia", ncbi_taxo = FALSE)
Aves_bold <- refdb::refdb_import_BOLD(taxon = "Aves", ncbi_taxo = FALSE)
Reptilia_bold <- refdb::refdb_import_BOLD(taxon = "Reptilia", ncbi_taxo = FALSE)
Actinopterygii_bold <- refdb::refdb_import_BOLD(taxon = "Actinopterygii", ncbi_taxo = FALSE)
Sarcopterygii_bold <- refdb::refdb_import_BOLD(taxon = "Sarcopterygii", ncbi_taxo = FALSE)

# Download for Proteobacteria for detection (and exclusion) of Wolbachia sequences
Proteobacteria_bold <- refdb::refdb_import_BOLD(taxon = "Proteobacteria", ncbi_taxo = FALSE)

# Download for plants for detection (and exclusion)
Bryophyta_bold <- refdb::refdb_import_BOLD(taxon = "Bryophyta", ncbi_taxo = FALSE)
Chlorophyta_bold <- refdb::refdb_import_BOLD(taxon = "Chlorophyta", ncbi_taxo = FALSE)
Lycopodiophyta_bold <- refdb::refdb_import_BOLD(taxon = "Lycopodiophyta", ncbi_taxo = FALSE)
Magnoliophyta_bold <- refdb::refdb_import_BOLD(taxon = "Magnoliophyta", ncbi_taxo = FALSE)
Pinophyta_bold <- refdb::refdb_import_BOLD(taxon = "Pinophyta", ncbi_taxo = FALSE)
Pteridophyta_bold <- refdb::refdb_import_BOLD(taxon = "Pteridophyta", ncbi_taxo = FALSE)

# Download for fungi for detection (and exclusion)
Ascomycota_bold <- refdb::refdb_import_BOLD(taxon = "Ascomycota", ncbi_taxo = FALSE)
Basidiomycota_bold <- refdb::refdb_import_BOLD(taxon = "Basidiomycota", ncbi_taxo = FALSE)
Chytridiomycota_bold <- refdb::refdb_import_BOLD(taxon = "Chytridiomycota", ncbi_taxo = FALSE)
Glomeromycota_bold <- refdb::refdb_import_BOLD(taxon = "Glomeromycota", ncbi_taxo = FALSE)
Myxomycota_bold <- refdb::refdb_import_BOLD(taxon = "Myxomycota", ncbi_taxo = FALSE)
Zygomycota_bold <- refdb::refdb_import_BOLD(taxon = "Zygomycota", ncbi_taxo = FALSE)
```

# get fields for download BOLD files
```{r}
refdb::refdb_get_fields(Annelida_bold)
refdb::refdb_get_fields(Merostomata_bold)
refdb::refdb_get_fields(Pauropoda_bold)
refdb::refdb_get_fields(Symphyla_bold)
refdb::refdb_get_fields(Diplopoda_bold)
refdb::refdb_get_fields(Chilopoda_bold)
refdb::refdb_get_fields(Copepoda_bold)
refdb::refdb_get_fields(Collembola_bold)
refdb::refdb_get_fields(Protura_bold)
refdb::refdb_get_fields(Cephalocarida_bold)
refdb::refdb_get_fields(Remipedia_bold)
refdb::refdb_get_fields(Branchiopoda_bold)
refdb::refdb_get_fields(Thecostraca_bold)
refdb::refdb_get_fields(Malacostraca_bold)
refdb::refdb_get_fields(Ichthyostraca_bold)
refdb::refdb_get_fields(Ostracoda_bold)
refdb::refdb_get_fields(Arachnida_bold)
refdb::refdb_get_fields(Diplura_bold)
refdb::refdb_get_fields(Pycnogonida_bold)
refdb::refdb_get_fields(Annelida_bold)
refdb::refdb_get_fields(Cnidaria_bold)
refdb::refdb_get_fields(Bryozoa_bold)
refdb::refdb_get_fields(Platyhelminthes_bold)
refdb::refdb_get_fields(Porifera_bold)
refdb::refdb_get_fields(Nematoda_bold)
refdb::refdb_get_fields(Rotifera_bold)
refdb::refdb_get_fields(Nemertea_bold)
refdb::refdb_get_fields(Onychophora_bold)
refdb::refdb_get_fields(Priapulida_bold)
refdb::refdb_get_fields(Tardigrada_bold)
refdb::refdb_get_fields(Mollusca_bold)
refdb::refdb_get_fields(Amphibia_bold)
refdb::refdb_get_fields(Leptocardii_bold)
refdb::refdb_get_fields(Mammalia_bold)
refdb::refdb_get_fields(Aves_bold)
refdb::refdb_get_fields(Reptilia_bold)
refdb::refdb_get_fields(Actinopterygii_bold)
refdb::refdb_get_fields(Sarcopterygii_bold)
refdb::refdb_get_fields(Archaeognatha_bold)
refdb::refdb_get_fields(Blattodea_bold)
refdb::refdb_get_fields(Coleoptera_bold)
refdb::refdb_get_fields(Dermaptera_bold)
refdb::refdb_get_fields(Embioptera_bold)
refdb::refdb_get_fields(Ephemeroptera_bold)
refdb::refdb_get_fields(Hemiptera_bold)
refdb::refdb_get_fields(Hymenoptera_bold)
refdb::refdb_get_fields(Lepidoptera_bold)
refdb::refdb_get_fields(Mantodea_bold)
refdb::refdb_get_fields(Mecoptera_bold)
refdb::refdb_get_fields(Megaloptera_bold)
refdb::refdb_get_fields(Neuroptera_bold)
refdb::refdb_get_fields(Notoptera_bold)
refdb::refdb_get_fields(Odonata_bold)
refdb::refdb_get_fields(Orthoptera_bold)
refdb::refdb_get_fields(Phasmatodea_bold)
refdb::refdb_get_fields(Plecoptera_bold)
refdb::refdb_get_fields(Psocodea_bold)
refdb::refdb_get_fields(Raphidioptera_bold)
refdb::refdb_get_fields(Siphonaptera_bold)
refdb::refdb_get_fields(Strepsiptera_bold)
refdb::refdb_get_fields(Thysanoptera_bold)
refdb::refdb_get_fields(Trichoptera_bold)
refdb::refdb_get_fields(Zoraptera_bold)
refdb::refdb_get_fields(Zygentoma_bold)
refdb::refdb_get_fields(Acartophthalmidae_bold)
refdb::refdb_get_fields(Acroceridae_bold)
refdb::refdb_get_fields(Agromyzidae_bold)
refdb::refdb_get_fields(Anisopodidae_bold)
refdb::refdb_get_fields(Anthomyiidae_bold)
refdb::refdb_get_fields(Anthomyzidae_bold)
refdb::refdb_get_fields(Apioceridae_bold)
refdb::refdb_get_fields(Apsilocephalidae_bold)
refdb::refdb_get_fields(Apystomyiidae_bold)
refdb::refdb_get_fields(Asilidae_bold)
refdb::refdb_get_fields(Asteiidae_bold)
refdb::refdb_get_fields(Atelestidae_bold)
refdb::refdb_get_fields(Athericidae_bold)
refdb::refdb_get_fields(Aulacigastridae_bold)
refdb::refdb_get_fields(Australimyzidae_bold)
refdb::refdb_get_fields(Austroleptidae_bold)
refdb::refdb_get_fields(Bibionidae_bold)
refdb::refdb_get_fields(Blephariceridae_bold)
refdb::refdb_get_fields(Bolbomyiidae_bold)
refdb::refdb_get_fields(Bolitophilidae_bold)
refdb::refdb_get_fields(Bombyliidae_bold)
refdb::refdb_get_fields(Calliphoridae_bold)
refdb::refdb_get_fields(Camillidae_bold)
refdb::refdb_get_fields(Canacidae_bold)
refdb::refdb_get_fields(Canthyloscelidae_bold)
refdb::refdb_get_fields(Carnidae_bold)
refdb::refdb_get_fields(Cecidomyiidae_bold)
refdb::refdb_get_fields(Celyphidae_bold)
refdb::refdb_get_fields(Ceratopogonidae_bold)
refdb::refdb_get_fields(Chamaemyiidae_bold)
refdb::refdb_get_fields(Chaoboridae_bold)
refdb::refdb_get_fields(Chironomidae_bold)
refdb::refdb_get_fields(Chloropidae_bold)
refdb::refdb_get_fields(Chyromyidae_bold)
refdb::refdb_get_fields(Clusiidae_bold)
refdb::refdb_get_fields(Coelopidae_bold)
refdb::refdb_get_fields(Conopidae_bold)
refdb::refdb_get_fields(Corethrellidae_bold)
refdb::refdb_get_fields(Cryptochetidae_bold)
refdb::refdb_get_fields(Ctenostylidae_bold)
refdb::refdb_get_fields(Culicidae_bold)
refdb::refdb_get_fields(Curtonotidae_bold)
refdb::refdb_get_fields(Cylindrotomidae_bold)
refdb::refdb_get_fields(Cypselosomatidae_bold)
refdb::refdb_get_fields(Deuterophlebiidae_bold)
refdb::refdb_get_fields(Diadocidiidae_bold)
refdb::refdb_get_fields(Diastatidae_bold)
refdb::refdb_get_fields(Diopsidae_bold)
refdb::refdb_get_fields(Diptera_family_incertae_sedis_bold)
refdb::refdb_get_fields(Ditomyiidae_bold)
refdb::refdb_get_fields(Dixidae_bold)
refdb::refdb_get_fields(Dolichopodidae_bold)
refdb::refdb_get_fields(Drosophilidae_bold)
refdb::refdb_get_fields(Dryomyzidae_bold)
refdb::refdb_get_fields(Empididae_bold)
refdb::refdb_get_fields(Ephydridae_bold)
refdb::refdb_get_fields(Evocoidae_bold)
refdb::refdb_get_fields(Fanniidae_bold)
refdb::refdb_get_fields(Fergusoninidae_bold)
refdb::refdb_get_fields(Glossinidae_bold)
refdb::refdb_get_fields(Gobryidae_bold)
refdb::refdb_get_fields(Helcomyzidae_bold)
refdb::refdb_get_fields(Heleomyzidae_bold)
refdb::refdb_get_fields(Helosciomyzidae_bold)
refdb::refdb_get_fields(Hesperinidae_bold)
refdb::refdb_get_fields(Hilarimorphidae_bold)
refdb::refdb_get_fields(Hippoboscidae_bold)
refdb::refdb_get_fields(Homalocnemidae_bold)
refdb::refdb_get_fields(Huttoninidae_bold)
refdb::refdb_get_fields(Hybotidae_bold)
refdb::refdb_get_fields(Inbiomyiidae_bold)
refdb::refdb_get_fields(Ironomyiidae_bold)
refdb::refdb_get_fields(Iteaphilidae_bold)
refdb::refdb_get_fields(Keroplatidae_bold)
refdb::refdb_get_fields(Lauxaniidae_bold)
refdb::refdb_get_fields(Limoniidae_bold)
refdb::refdb_get_fields(Lonchaeidae_bold)
refdb::refdb_get_fields(Lonchopteridae_bold)
refdb::refdb_get_fields(Marginidae_bold)
refdb::refdb_get_fields(Megamerinidae_bold)
refdb::refdb_get_fields(Mesembrinellidae_bold)
refdb::refdb_get_fields(Micropezidae_bold)
refdb::refdb_get_fields(Milichiidae_bold)
refdb::refdb_get_fields(Muscidae_bold)
refdb::refdb_get_fields(Mycetophilidae_bold)
refdb::refdb_get_fields(Mydidae_bold)
refdb::refdb_get_fields(Mystacinobiidae_bold)
refdb::refdb_get_fields(Nannodastiidae_bold)
refdb::refdb_get_fields(Natalimyzidae_bold)
refdb::refdb_get_fields(Nemestrinidae_bold)
refdb::refdb_get_fields(Neriidae_bold)
refdb::refdb_get_fields(Nothybidae_bold)
refdb::refdb_get_fields(Nymphomyiidae_bold)
refdb::refdb_get_fields(Odiniidae_bold)
refdb::refdb_get_fields(Oestridae_bold)
refdb::refdb_get_fields(Opetiidae_bold)
refdb::refdb_get_fields(Opomyzidae_bold)
refdb::refdb_get_fields(Oreogetonidae_bold)
refdb::refdb_get_fields(Pachyneuridae_bold)
refdb::refdb_get_fields(Pallopteridae_bold)
refdb::refdb_get_fields(Pantophthalmidae_bold)
refdb::refdb_get_fields(Paraleucopidae_bold)
refdb::refdb_get_fields(Pediciidae_bold)
refdb::refdb_get_fields(Pelecorhynchidae_bold)
refdb::refdb_get_fields(Periscelididae_bold)
refdb::refdb_get_fields(Phoridae_bold)
refdb::refdb_get_fields(Piophilidae_bold)
refdb::refdb_get_fields(Pipunculidae_bold)
refdb::refdb_get_fields(Platypezidae_bold)
refdb::refdb_get_fields(Platystomatidae_bold)
refdb::refdb_get_fields(Polleniidae_bold)
refdb::refdb_get_fields(Pseudopomyzidae_bold)
refdb::refdb_get_fields(Psilidae_bold)
refdb::refdb_get_fields(Psychodidae_bold)
refdb::refdb_get_fields(Ptychopteridae_bold)
refdb::refdb_get_fields(Pyrgotidae_bold)
refdb::refdb_get_fields(Rhagionidae_bold)
refdb::refdb_get_fields(Richardiidae_bold)
refdb::refdb_get_fields(Ropalomeridae_bold)
refdb::refdb_get_fields(Sarcophagidae_bold)
refdb::refdb_get_fields(Scathophagidae_bold)
refdb::refdb_get_fields(Scatopsidae_bold)
refdb::refdb_get_fields(Scenopinidae_bold)
refdb::refdb_get_fields(Sciaridae_bold)
refdb::refdb_get_fields(Sciaroidea_incertae_sedis_bold)
refdb::refdb_get_fields(Sciomyzidae_bold)
refdb::refdb_get_fields(Sepsidae_bold)
refdb::refdb_get_fields(Simuliidae_bold)
refdb::refdb_get_fields(Somatiidae_bold)
refdb::refdb_get_fields(Sphaeroceridae_bold)
refdb::refdb_get_fields(Stratiomyidae_bold)
refdb::refdb_get_fields(Strongylophthalmyiidae_bold)
refdb::refdb_get_fields(Syringogastridae_bold)
refdb::refdb_get_fields(Syrphidae_bold)
refdb::refdb_get_fields(Tabanidae_bold)
refdb::refdb_get_fields(Tachinidae_bold)
refdb::refdb_get_fields(Tanyderidae_bold)
refdb::refdb_get_fields(Tanypezidae_bold)
refdb::refdb_get_fields(Tephritidae_bold)
refdb::refdb_get_fields(Thaumaleidae_bold)
refdb::refdb_get_fields(Therevidae_bold)
refdb::refdb_get_fields(Tipulidae_bold)
refdb::refdb_get_fields(Trichoceridae_bold)
refdb::refdb_get_fields(Ulidiidae_bold)
refdb::refdb_get_fields(Vermileonidae_bold)
refdb::refdb_get_fields(Xylomyidae_bold)
refdb::refdb_get_fields(Xylophagidae_bold)
refdb::refdb_get_fields(Proteobacteria_bold)
refdb::refdb_get_fields(Bryophyta_bold)
refdb::refdb_get_fields(Chlorophyta_bold)
refdb::refdb_get_fields(Lycopodiophyta_bold)
refdb::refdb_get_fields(Magnoliophyta_bold)
refdb::refdb_get_fields(Pinophyta_bold)
refdb::refdb_get_fields(Pteridophyta_bold)
refdb::refdb_get_fields(Ascomycota_bold)
refdb::refdb_get_fields(Basidiomycota_bold)
refdb::refdb_get_fields(Chytridiomycota_bold)
refdb::refdb_get_fields(Glomeromycota_bold)
refdb::refdb_get_fields(Myxomycota_bold)
refdb::refdb_get_fields(Zygomycota_bold)
```

There is an issue here. I can not merge the BOLD dataset together as the lat and lon variables in some files are character and some numeric. This still needs work.
##Chris's suggestions to fix lat long inconsistency in vector type (for next download)
# inspect what might be going wrong with unique
unique(x$lat)
unique(x$lonh)
# maybe this will work if R thinks x$latitude is a character rather than a number
x$lat <- as.numeric(x$lat)

To work around this I have used the control spike sample of Scaptodrosophila that does not contain lat and long data to merge data and 'keep field shared'. The lat and long data is not needed for database construction.

# load control spike sample
```{r}
# load control sample
control <- read.csv(here::here("raw_libraries/control_spike.csv"))

# set fields for refdb
control <- refdb::refdb_set_fields(control,
                        taxonomy = c(
                        phylum = "phylum",
                        class = "class",
                        order = "order",
                        family = "family",
                        genus = "genus",
                        species = "species"),
                        sequence = "DNA_seq",
                        marker = "marker",
                        source = "source",
                        id = "id")

refdb::refdb_get_fields(control)
```

#Step 3: Merging all BOLD animal data into one dataframe
```{r}
bold_euk_data <- refdb::refdb_merge(Annelida_bold, Cnidaria_bold, Bryozoa_bold, Platyhelminthes_bold, Porifera_bold, Nematoda_bold, Rotifera_bold, Nemertea_bold, Onychophora_bold, Priapulida_bold, Tardigrada_bold, Mollusca_bold, Merostomata_bold, Pauropoda_bold, Symphyla_bold, Diplopoda_bold, Chilopoda_bold, Copepoda_bold, Collembola_bold, Protura_bold, Cephalocarida_bold, Remipedia_bold, Branchiopoda_bold, Thecostraca_bold, Malacostraca_bold, Ichthyostraca_bold, Diplura_bold, Pycnogonida_bold, Ostracoda_bold, Arachnida_bold, Archaeognatha_bold, Blattodea_bold, Coleoptera_bold, Dermaptera_bold, Embioptera_bold, Ephemeroptera_bold, Hemiptera_bold, Hymenoptera_bold, Lepidoptera_bold, Mantodea_bold, Mecoptera_bold, Megaloptera_bold, Neuroptera_bold, Notoptera_bold, Odonata_bold, Orthoptera_bold, Phasmatodea_bold, Plecoptera_bold, Psocodea_bold, Raphidioptera_bold, Siphonaptera_bold, Strepsiptera_bold, Thysanoptera_bold, Trichoptera_bold, Zoraptera_bold, Zygentoma_bold, Acartophthalmidae_bold, Acroceridae_bold, Agromyzidae_bold, Anisopodidae_bold, Anthomyiidae_bold, Anthomyzidae_bold, Apioceridae_bold, Apsilocephalidae_bold, Apystomyiidae_bold, Asilidae_bold, Asteiidae_bold, Atelestidae_bold, Athericidae_bold, Aulacigastridae_bold, Australimyzidae_bold, Austroleptidae_bold, Bibionidae_bold, Blephariceridae_bold, Bolbomyiidae_bold, Bolitophilidae_bold, Bombyliidae_bold, Calliphoridae_bold, Camillidae_bold, Canacidae_bold, Canthyloscelidae_bold, Carnidae_bold, Cecidomyiidae_bold, Celyphidae_bold, Ceratopogonidae_bold, Chamaemyiidae_bold, Chaoboridae_bold, Chironomidae_bold, Chloropidae_bold, Chyromyidae_bold, Clusiidae_bold, Coelopidae_bold, Conopidae_bold, Corethrellidae_bold, Cryptochetidae_bold, Ctenostylidae_bold, Culicidae_bold, Curtonotidae_bold, Cylindrotomidae_bold, Cypselosomatidae_bold, Deuterophlebiidae_bold, Diadocidiidae_bold, Diastatidae_bold, Diopsidae_bold, Diptera_family_incertae_sedis_bold, Ditomyiidae_bold, Dixidae_bold, Dolichopodidae_bold, Drosophilidae_bold, Dryomyzidae_bold, Empididae_bold, Ephydridae_bold, Evocoidae_bold, Fanniidae_bold, Fergusoninidae_bold, Glossinidae_bold, Gobryidae_bold, Helcomyzidae_bold, Heleomyzidae_bold, Helosciomyzidae_bold, Hesperinidae_bold, Hilarimorphidae_bold, Hippoboscidae_bold, Homalocnemidae_bold, Huttoninidae_bold, Hybotidae_bold, Inbiomyiidae_bold, Ironomyiidae_bold, Iteaphilidae_bold, Keroplatidae_bold, Lauxaniidae_bold, Limoniidae_bold, Lonchaeidae_bold, Lonchopteridae_bold, Marginidae_bold, Megamerinidae_bold, Mesembrinellidae_bold, Micropezidae_bold, Milichiidae_bold, Muscidae_bold, Mycetophilidae_bold, Mydidae_bold, Mystacinobiidae_bold, Nannodastiidae_bold, Natalimyzidae_bold, Nemestrinidae_bold, Neriidae_bold, Nothybidae_bold, Nymphomyiidae_bold, Odiniidae_bold, Oestridae_bold, Opetiidae_bold, Opomyzidae_bold, Oreogetonidae_bold, Pachyneuridae_bold, Pallopteridae_bold, Pantophthalmidae_bold, Paraleucopidae_bold, Pediciidae_bold, Pelecorhynchidae_bold, Periscelididae_bold, Phoridae_bold, Piophilidae_bold, Pipunculidae_bold, Platypezidae_bold, Platystomatidae_bold, Polleniidae_bold, Pseudopomyzidae_bold, Psilidae_bold, Psychodidae_bold, Ptychopteridae_bold, Pyrgotidae_bold, Rhagionidae_bold, Richardiidae_bold, Ropalomeridae_bold, Sarcophagidae_bold, Scathophagidae_bold, Scatopsidae_bold, Scenopinidae_bold, Sciaridae_bold, Sciaroidea_incertae_sedis_bold, Sciomyzidae_bold, Sepsidae_bold, Simuliidae_bold, Somatiidae_bold, Sphaeroceridae_bold, Stratiomyidae_bold, Strongylophthalmyiidae_bold, Syringogastridae_bold, Syrphidae_bold, Tabanidae_bold, Tachinidae_bold, Tanyderidae_bold, Tanypezidae_bold, Tephritidae_bold, Thaumaleidae_bold, Therevidae_bold, Tipulidae_bold, Trichoceridae_bold, Ulidiidae_bold, Vermileonidae_bold, Xylomyidae_bold, Xylophagidae_bold, Amphibia_bold, Leptocardii_bold, Mammalia_bold, Aves_bold, Reptilia_bold, Actinopterygii_bold, Sarcopterygii_bold, control, keep = "fields_shared")
```

# final formating of dataframe - animals
```{r}
# add superkingdom column to bring in line with NCBI GenBank data
bold_euk_data <- bold_euk_data %>%
dplyr::mutate(skingdom_name = "Eukaryota",
  .before = "phylum_name")

# set fields for refdb
bold_euk_data <- refdb::refdb_set_fields(bold_euk_data,
                             source = "source",
                             id = "sequenceID",
                             taxonomy = c(kingdom = "superkingdom_name",
                                          phylum = "phylum_name",
                                          class = "class_name",
                                          order = "order_name",
                                          family = "family_name",
                                          genus = "genus_name",
                                          species = "species_name"),
                             sequence = "nucleotides",
                             marker = "markercode")

```

# final formating of dataframe - Proteobacteria
```{r}
# add superkingdom column to bring in line with NCBI GenBank data
bold_prot_data <- Proteobacteria_bold %>%
dplyr::mutate(superkingdom_name = "Prokaryota",
  .before = "phylum_name")

# set fields for refdb
bold_prot_data <- refdb::refdb_set_fields(bold_prot_data,
                             source = "source",
                             id = "sequenceID",
                             taxonomy = c(kingdom = "superkingdom_name",
                                          phylum = "phylum_name",
                                          class = "class_name",
                                          order = "order_name",
                                          family = "family_name",
                                          genus = "genus_name",
                                          species = "species_name"),
                             sequence = "nucleotides",
                             marker = "markercode")

```

# final formating of dataframe - Plants and fungi
```{r}
# fix and lat or long values that are not numeric
Magnoliophyta$lat <- as.numeric(Magnoliophyta$lat)

# merge plant and fungal sequences
bold_plafun_data <- refdb::refdb_merge(Bryophyta_bold, Chlorophyta_bold, Lycopodiophyta_bold, Magnoliophyta_bold, Pinophyta_bold, Pteridophyta_bold, Ascomycota_bold, Basidiomycota_bold, Chytridiomycota_bold, Glomeromycota_bold, Myxomycota_bold,Zygomycota_bold)

# add superkingdom column to bring in line with NCBI GenBank data
bold_plafun_data <- bold_plafun_data %>% 
dplyr::mutate(superkingdom_name = "Eukaryota",
  .before = "phylum_name")

# set fields for refdb
bold_plafun_data <- refdb::refdb_set_fields(bold_plafun_data,
                             source = "source",
                             id = "sequenceID",
                             taxonomy = c(kingdom = "superkingdom_name",
                                          phylum = "phylum_name",
                                          class = "class_name",
                                          order = "order_name",
                                          family = "family_name",
                                          genus = "genus_name",
                                          species = "species_name"),
                             sequence = "nucleotides",
                             marker = "markercode")
                                    
```

# merge all bold data into a single file
```{r}
bold_data <- refdb::refdb_merge(bold_prot_data, bold_euk_data, bold_plafun_data)
```

# Step 5: save merged raw data from BOLD library (note change date for new downloads when saving)
```{r}
write_rds(bold_data, (here::here("/refdb_input_data/refdb_input_data/bold_data_oct23.rds"))

# reload all BOLD document for cleaning (if needed)
bold_data <- readRDS(here::here("refdb_input_data/bold_data_oct23.rds"))

```


```{r}
# setting fields for all BOLD data
bold_data <- refdb::refdb_set_fields(bold_data,
                             source = "source",
                             id = "sequenceID",
                             taxonomy = c(kingdom = "superkingdom_name",
                                          phylum = "phylum_name",
                                          class = "class_name",
                                          order = "order_name",
                                          family = "family_name",
                                          genus = "genus_name",
                                          species = "species_name"),
                             sequence = "nucleotides",
                             marker = "markercode")
```

## Preparing and cleaning DNA sequences and taxonomic data from BOLD

#Step 6: Pulling out COI sequences and cleaning of data
```{r}
#Filter for CO1 only. BOLD contains sequences mostly COI but also DNA sequences from other gene regions.The scripts commented out below can be used to group sequences on BOLD for difference gene regions

bold_data_COI <- subset(bold_data, markercode=="COI-5P")
#bold_data_18S <- subset(bold_data, markercode=="18S")
#bold_data_ITS <- subset(bold_data, markercode=="ITS")
#bold_data_12S <- subset(bold_data, markercode=="12S")
#bold_data_ND2 <- subset(bold_data, markercode=="ND2")
#bold_data_16S <- subset(bold_data, markercode=="16S")

# Remove sequence alignment gaps from BOLD library:
bold_data_clean1 <- refdb::refdb_clean_seq_remove_gaps(bold_data_COI)

# Remove side N's (removes N's at start or end of sequence)
bold_data_clean2 <- refdb::refdb_clean_seq_remove_sideN(bold_data_clean1)

# Remove extra bits on the end of species names
bold_data_clean <- refdb::refdb_clean_tax_remove_extra(bold_data_clean2)

#remove files no longer needed
rm(bold_data_clean1)
rm(bold_data_clean2)
rm(bold_data)
rm(bold_data_COI)
```

Note: DO NOT use remove refdb_clean_tax_NA. There is a bug as it removes interim names and the letters 'na'  within an taxonomic names. Another script is provided to remove 'NA' from the library so that 'NA's don't appear in taxonomic information of the final library files.

#Step 7: Remove duplicates. It is good to write this file prior to preforming other clean_up steps
This step removes a large amount of data it looks for records with identical sequences and taxonomic information and removes all duplication. Expect the number of records to reduce by 30-50%.
```{r}
bold_data_clean <- refdb::refdb_filter_seq_duplicates(bold_data_clean)
```

#Step 8:Plot sequence length
Gives an idea of the distribution of the data
```{r}
refdb::refdb_plot_seqlen_hist(bold_data_clean)
```

#Visualizing the reference library (Optional)
Some of the functions you can use to produce graphical representation of your reference database. Because refdb stores reference database as dataframes it is straightforward to produce plots (e.g. with the tidyverse and ggplot2). For example, we can make a barplot showing the distribution of the families like this:

```{r}
bold_data_clean %>% 
  group_by(phylum_name) %>% 
  count() %>% 
  ggplot2::ggplot(aes(fct_reorder(phylum_name, n, .desc = TRUE), n)) +
  geom_col() +
  xlab("Phylum") +
  ylab("Number of records") +
  theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust=1))
```

```{r}
bold_data_clean %>% 
  group_by(class_name) %>% 
  count() %>% 
  ggplot2::ggplot(aes(fct_reorder(class_name, n, .desc = TRUE), n)) +
  geom_col() +
  xlab("Class") +
  ylab("Number of records") +
  theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust=1))
```


```{r}
refdb::refdb_plot_tax_treemap(bold_data_clean, cols = 2)
```

Alternatively you can represent this information with a taxonomic tree. These functions have several parameters to control what is represented in the plot (taxonomic ranks, colors, etc.).

```{r}
refdb::refdb_plot_tax_tree(bold_data_clean)
```

#creating a report (worth doing)
A report is a simple and rapid way to get an overview of the current state of your reference library and to identify some possible issues. You can compile a report using the function refdb_report: 
  
```{r}
refdb::refdb_report(bold_data_clean, view=TRUE)
```                       

# Step 9: Export cleaned data (add database, datatype and date)
This data will be used to create the DNA barcode reference library and will be combined with the data  below which includes downloads from NCBI GenBank and data from our own DNA sequencing (private DNA barcodes)

```{r}
# save cleaned data
saveRDS(bold_data_clean, here::here("cleaned_libraries/bold_data_clean_Oct23.rds"))

# import cleaned BOLD data (if required)
#bold_data_clean <- readRDS(here::here("cleaned_libraries/bold_data_clean_Oct23.rds"))
```

--------------------------------------------------------------------------------
## SECTION 3: Preparing a private library as a DNA reference database

This step imports DNA barcodes which I have produced from invertebrates. It is a mixture of unpublished and published data. The private reference library is managed in Geneious prime and can be exported out as .csv. Private libraries can be imported into this documant as .csv files and need to contain at a minimum: The lowest possible taxonomic classification (superkingdom;phylum;class;order;family;genus;species), an individual sequence identifier (i.e., sequence, sample or voucher code) and a COI DNA sequence. Other metadata can be included but will not be used for reference library construction.

Please refer to the private library checking scripts (in 'Checking private library data' workflowvdocument) before adding private sequence data. This document can be used to vet data for errors

# read in DNA barcode datafile in .csv 
```{r}
priv_lib <- read_csv (here::here("raw_libraries/priv_lib_Nov23.csv"))
```

# set fields for refdb
```{r}
priv_lib <- refdb::refdb_set_fields(priv_lib,
                             taxonomy = c(kingdom ="superkingdom",
                               phylum = "phylum",
                               class = "class",
                               order = "order",
                               family = "family",
                               genus = "genus",
                               species = "species"),
                             sequence = "DNA_seq",
                             marker = "marker",
                             source = "source",
                             id = "id")

refdb::refdb_get_fields(priv_lib)
```
#Step 1: Remove duplicates.
```{r}
priv_lib_clean <- refdb::refdb_filter_seq_duplicates(priv_lib)

rm(priv_lib)
```

# plot sequence length
```{r}
refdb::refdb_plot_seqlen_hist(priv_lib_clean)
```

Note: Primer cropping in refdb does not work! I could not get refdb to properly trim primers it needed really high error setting to do any trimming, but with these settings over trimmed most of the data. for the short amplicon the forward primer trims ok but the reverse trims in the wrong spot.

#Visualizing the reference library (Optional)

Now letâ€™s take a tour of the functions you can use to produce graphical representation of your reference database. Because refdb stores reference database as dataframes it is straightforward to produce plots (e.g. with the tidyverse and ggplot2). For example, we can make a barplot showing the distribution of the families like this:
  
```{r}
priv_lib_clean %>% 
  group_by(phylum) %>% 
  count() %>% 
  ggplot2::ggplot(aes(fct_reorder(phylum, n, .desc = TRUE), n)) +
  geom_col() +
  xlab("Phylum") +
  ylab("Number of records") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
```

```{r}
priv_lib_clean %>% 
  group_by(class) %>% 
  count() %>% 
  ggplot(aes(fct_reorder(class, n, .desc = TRUE), n)) +
  geom_col() +
  xlab("Class") +
  ylab("Number of records") +
  theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust=1))
```

Additionally, refdb provides ready-to-use functions to produce more sophisticated plots. For example, to represent the taxonomic coverage of your reference library across multiple levels with a tree map:
  
Jess - This does not work but not important as there is simply too much data to get anything meaningful out.

```{r}
refdb::refdb_plot_tax_treemap(priv_lib_clean, cols = 4)
```

Alternatively you can represent this information with a taxonomic tree. These functions have several parameters to control what is represented in the plot (taxonomic ranks, colors, etc.).

```{r}
refdb::refdb_plot_tax_tree(priv_lib_clean)
```

# Creating a report
A report is a simple and rapid way to get an overview of the current state of your reference library and to identify some possible issues. You can compile a report using the function refdb_report: 
  
```{r}
refdb::refdb::refdb_report(priv_lib_clean, view=TRUE)
```                    

# Exporting cleaned private data
Once satisfied with the results, libraries can be exported for further use. Since the reference database is a dataframe, it is possible to use any export functions (from R or from other packages) that support dataframes.

```{r}
# importing library into R (if required)
saveRDS(priv_lib_clean,(here::here("cleaned_libraries/priv_lib_clean_Oct23.rds")))

# reload all BOLD document for cleaning (if needed)
#priv_lib_clean <- readRDS(here::here("cleaned_libraries/bold_data_oct23.rds"))

```

# set fields for priv_lib_clean
```{r}
priv_lib_clean <- refdb::refdb_set_fields(priv_lib_clean,
                             taxonomy = c(kingdom = "superkingdom",
                               phylum = "phylum",
                               class = "class",
                               order = "order",
                               family = "family",
                               genus = "genus",
                               species = "species"),
                             sequence = "DNA_seq",
                             marker = "marker",
                             source = "source",
                             id = "id")

refdb::refdb_get_fields(priv_lib_clean)
```

--------------------------------------------------------------------------------
# SECTION 4: Creating a complete reference library

# Step 1:  Merging data from different sources
We can now use refdb_merge to merge all data into a single database.

# mergeing all libraries into a single document
```{r}
all_bold_priv_ncbi <- refdb::refdb_merge(priv_lib_clean, ncbi_data_clean, bold_data_clean, keep = "fields_shared")

rm(bold_data_clean)
rm(ncbi_data_clean)
rm(priv_lib_clean)
```

# adjust taxonomy across datasets
Ensures that the consist taxonomic classification is use in specified columns (Phyla to genus, leave out species as it removes interim names (specify columns). This should be used on the final dataset (with GB data included) for the taxonomic field of phylum, class, order, family, genus. Species need to be left out as this function will remove interim species names from the private database.

```{r}
all_bold_priv_ncbi <- refdb::refdb_clean_tax_harmonize_nomenclature(all_bold_priv_ncbi, 4:7)
head(all_bold_priv_ncbi)
```

# remove duplicates between private, BOLD and NCBI GenBank datasets
```{r}
all_bold_priv_ncbi <- refdb::refdb_filter_seq_duplicates(all_bold_priv_ncbi)
```

# resolving issues from refdb in the final library
The final library will still have NA in some empty taxonomic columns as the refdb function removes 'na' within taxonomic names. The final .rds will need the NA's filtered out using code that only removes 'NA from entire cells.
Also, there are some strange additions to the end for species names from the 'remove extra' that should be removed these include:

# Step 1: remove 'NA' from empty cells, weird additions on species names and trim whitespaces remaining at the end of species names
```{r}
#remove "NA" in taxonomic information and replace with blank (this needs to be done so NA do not occur in the final reference library taxonomic assignments)

all_bold_priv_ncbi[is.na(all_bold_priv_ncbi)] <- ""

#replacing odd additions to the end of species names from refdb

all_bold_priv_ncbi$species <- gsub("\\'", "",
           gsub("\\.\\.", "",
           gsub("\\+", "",
           gsub("./Se/", "",
           gsub("sp. -", "sp. ",
           gsub("\\:", "",
           gsub(".cf.", "",
           gsub("sp.-", "sp. ",
           gsub("_", "",
           gsub("/Amp", "",
           gsub("<>", "",
           gsub("\\  ", "", all_bold_priv_ncbi$species))))))))))))

#remove trailing spaces in species names
trimws(all_bold_priv_ncbi$species, "right")

missing_rows <- which(is.na(all_bold_priv_ncbi$id))

print(missing_rows)
```

#Step 2: Fix outdated taxonomic names 
```{r}
# fix outdated taxonomic names (Changing Ancylidae to Planorbidae, Sphaeriidae to Pisidiidae, Corbiculidae to Cyrenidae, Brentidae to Nanophyidae)

all_bold_priv_ncbi$family <- gsub("Ancylidae",
           "Planorbidae",
           gsub("Sphaeriidae",
           "Pisidiidae",
           gsub("Corbiculidae",
           "Cyrenidae",
           gsub("Brentidae",
           "Nanophyidae", all_bold_priv_ncbi$family))))

# check changes should all be TRUE
#print (sum(str_detect(all_bold_priv_NCBI$family, 'Ancylidae')) > 0)
#print (sum(str_detect(all_bold_priv_NCBI$family, 'Sphaeriidae')) > 0)
#print (sum(str_detect(all_bold_priv_NCBI$family, 'Corbiculidae')) > 0)
#print (sum(str_detect(all_bold_priv_NCBI$family, 'Brentidae')) > 0)
```  

# export final library as a dataframe
```{r}
saveRDS(all_bold_priv_ncbi, here::here("final_library/all_bold_priv_NCBI_final_clean_Nov23.rds"))

#read in if needed
all_bold_priv_ncbi <- readRDS(here::here("final_library/all_bold_priv_NCBI_final_clean_Nov23.rds"))
```

# convert to fasta file for import to Geneious
```{r}
# files are too large to export in one go so need to be broken in to smaller parts.

# create a duplicate document to format for Geneious 
all_bold_priv_ncbi2 <- all_bold_priv_ncbi

# add a an underscore to species to replace space in species names (fasta files cannot have gaps)
all_bold_priv_ncbi2$species <- gsub(" ",
           "_", all_bold_priv_ncbi$species)

head(all_bold_priv_ncbi2)

# export group 1
export <- all_bold_priv_ncbi2[1:2000000, ] 

# save fasta 
refdb_export_dada2(export, here::here("geneious_library/final_lib_export_nov23_1of3.fasta", mode = "taxonomy"))

# remove files no longer needed
rm(export)

# export group 2
export2 <- all_bold_priv_ncbi2[2000001:4000000, ] 

# save fasta 
refdb_export_dada2(export2, here::here("geneious_library/final_lib_export_nov23_2of3.fasta", mode = "taxonomy"))

# remove files no longer needed
rm(export2)

# export group 3 (adjust final row number to match the final record number)
export3 <- all_bold_priv_ncbi2[4000000:5520322, ] 

# save fasta 
refdb_export_dada2(export3, here::here("geneious_library/final_lib_export_nov23_3of3.fasta", mode = "taxonomy"))

# remove files no longer needed
rm(export3)
rm(all_bold_priv_ncbi2)

# trialing ways to combine split fasta files. This can be done outside of r using Geneious
#fasta1 <- bioseq::read_fasta(here::here("geneious_library/final_lib_export_1.fasta")

#fasta2 <- bioseq::read_fasta(here::here("geneious_library/final_lib_export_2.fasta")

#fasta3 <- bioseq::read_fasta(here::here("geneious_library/final_lib_export_3.fasta")
 
#comb_comp_lib <-  bioseq::seq_combine(fasta1, fasta2, fasta3)

```


--------------------------------------------------------------------------------
#SECTION 5: Creating amplicon specific libraries for bioinformatic pipelines
Unfortunately, the primer cropping in refdb does not work! I could not get refdb to properly trim primers it needed really high error setting to do any trimming, but with these settings over trimmed most of the data. for the short amplicon the forward primer trims ok but the reverse trims in the wrong spot.

Therefore for this step we can use cutadapt.....
# Load packages
```{r}
library(dplyr)
library(refdb)
library(bioseq)
library(stringr)
```

# load library
```{r}
# rename data for futher analysis
all_db <- all_bold_priv_ncbi

#all_db <-priv_lib_clean

# remove files no longer needed
rm(all_bold_priv_ncbi)

#change column header from 'DNA_seq' to 'sequence' for further analysis
colnames(all_db)[colnames(all_db) == "DNA_seq"] ="sequence"

#count rows
nrow(DNA_seq)
```
#Step 1: set fields for refdb
```{r}
all_db <- refdb::refdb_set_fields(all_db,
                             taxonomy = c(kingdom = "superkingdom",
                               phylum = "phylum",
                               class = "class",
                               order = "order",
                               family = "family",
                               genus = "genus",
                               species = "species"),
                             sequence = "sequence",
                             marker = "marker",
                             source = "source",
                             id = "id")

refdb::refdb_get_fields(all_db)
```

```{r}
# check there are no duplicate ids codes
length(unique(all_db$id)) == nrow(all_db)

#if true move on
```

```{r}
# check flie size
all_db %>% object.size()
```

# check file column names (DNA_seq = sequence data)
```{r}
head (all_db)
```

#Step 2: Convert to fasta file for trimming in cutadapt
The file size is large so it can be partioned for this step (in 4 parts here). The input files for cutadapt is .fasta
```{r}
seqs <- all_db[, "sequence",drop=TRUE]
names(seqs) <- 1:nrow(all_db)
parts <- as.numeric(cut(seq_along(seqs), 4))
bioseq::write_fasta(seqs[parts == 1], file = here::here("final_library/trimming/all_db.fasta"),
                    line_length = 80, block_length=Inf, append=FALSE)
bioseq::write_fasta(seqs[parts == 2], file = here::here("final_library/trimming/all_db.fasta"),
                    line_length = 80, block_length=Inf, append=TRUE)
bioseq::write_fasta(seqs[parts == 3], file = here::here("final_library/trimming/all_db.fasta"),
                    line_length = 80, block_length=Inf, append=TRUE)
bioseq::write_fasta(seqs[parts == 4], file = here::here("final_library/trimming/all_db.fasta"),
                    line_length = 80, block_length=Inf, append=TRUE)
```

```{r}
# Append qiime2 path to PATH
current_path_env <- Sys.getenv("PATH")

# Also set qiime2 conda environment in batchtools.slurm.tmpl
Sys.setenv(PATH=paste0("/mnt/galaxy/gvl/software/shared_envs/qiime2_2023-07/bin:",
                       current_path_env))
Sys.setenv(PYTHONPATH="/mnt/galaxy/gvl/software/shared_envs/qiime2_2023-07/lib/python3.8/site-packages/")
```
#Step 3: use Cutadapt (python) to create amplicon specific DNA barcode reference libraries.
If working on the ARC macroinvertebrate project the libraries will only need to be set up for the "short" and the "right" amplicon. 
```{bash, engine.opts='-l'}
# Mel_comment -doesn't seem to run as a bash chunk. I need to run the commands by copying and pasting into the terminal.

# # module load cutadapt/4.2 - first time only?
#conda create -n metabarcoding
#conda activate metabarcoding
#conda install -c bioconda cutadapt

#setting working directories doesn't work for Mel?

#cd ~/git/DNA_reference_db_construction/refdb/final_library/trim_seq

#check cutadapt is loaded in terminal cutadapt --version (give 3.4 if laoded)

## scripts for trimming for each amplicon (short, right, long and left)

cd ~/git/DNA_barcode_reference_database_construction/final_library

output_dir=$(pwd)

# strict trim
ampliconGroup=short    
cutadapt -g ACWGGWTGRACWGTNTAYCC...GCHGGDGCHATYACHATRYT -e 0.1 --revcomp -m 100 -M 450 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short1.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long1.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed1.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed1.fasta \
    ${output_dir}/all_db.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt.log
    
 # relax trim on long 
    ampliconGroup=short    
cutadapt -g ACWGGWTGRACWGTNTAYCC...GCHGGDGCHATYACHATRYT -e 0.31 --revcomp -m 100 -M 450 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short2.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long2.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed2.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed2.fasta \
    ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long1.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt2.log

 # relax trim on untrimmed    
cutadapt -g ACWGGWTGRACWGTNTAYCC...GCHGGDGCHATYACHATRYT -e 0.31 --revcomp -m 100 -M 450 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short3.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long3.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed3.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed3.fasta \
    ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed1.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt3.log
    
cat ${output_dir}/trimming/short_amplicon_trimmed1.fasta  ${output_dir}/trimming/short_amplicon_trimmed2.fasta  ${output_dir}/trimming/short_amplicon_trimmed3.fasta > ${output_dir}/short_amplicon_trimmed.fasta
```

#right amplicon
```{r}

#cd ~/git/DNA_barcode_reference_database_construction/final_library

# initial strict trim
bash_command <-
"conda activate metabarcoding
output_dir=$(pwd)
ampliconGroup=right
cutadapt -g AACWGGWTGRACWGTNTAYCC...TGRTTYTTYGGNCAYCCHGA -e 0.15 --revcomp -m 200 -M 550 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short1.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long1.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed1.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed1.fasta \
    ${output_dir}/all_db.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt1.log"
    
system(bash_command)
```    

# relax trim on too long 
ampliconGroup=right    
cutadapt -g ACWGGWTGRACWGTNTAYCC...TGRTTYTTYGGNCAYCCHGA -e 0.3 --revcomp -m 200 -M 550 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short2.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long2.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed2.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed2.fasta \
    ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long1.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt2.log

cutadapt -g ACWGGWTGRACWGTNTAYCC...TGRTTYTTYGGNCAYCCHGA -e 0.36 --revcomp -m 200 -M 550 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short4.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long4.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed4.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed4.fasta \
    ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long2.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt4.log

 # relax trim on untrimmed    
cutadapt -g ACWGGWTGRACWGTNTAYCC...TGRTTYTTYGGNCAYCCHGA -e 0.32 --revcomp -m 200 -M 550 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short3.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long3.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed3.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed3.fasta \
    ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed1.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt3.log
    
cat ${output_dir}/trimming/right_amplicon_trimmed1.fasta ${output_dir}/trimming/right_amplicon_trimmed2.fasta ${output_dir}/trimming/right_amplicon_trimmed3.fasta ${output_dir}/trimming/right_amplicon_trimmed4.fasta > ${output_dir}/right_amplicon_trimmed.fasta
 
cat  ~/git/DNA_barcode_reference_database_construction/final_library/trimming/right_amplicon_trimmed1.fasta  ~/git/DNA_barcode_reference_database_construction/final_library/trimming/right_amplicon_trimmed2.fasta  ~/git/DNA_barcode_reference_database_construction/final_library/trimming/right_amplicon_trimmed3.fasta  ~/git/DNA_barcode_reference_database_construction/final_library/trimming/right_amplicon_trimmed4.fasta > ~/git/DNA_barcode_reference_database_construction/final_library/right_amplicon_trimmed.fasta
    

# Join trimmed outputs back together into a single fasta file
```{r}
# join trimmed outputs
fasta_file1 <- readLines("/mnt/galaxy/home/mecarew/Documents/DNA_reference_db_construction/refdb/final_library/trimming/right_amplicon_trimmed1.fasta")
fasta_file2 <- readLines("/mnt/galaxy/home/mecarew/Documents/DNA_reference_db_construction/refdb/final_library/trimming/right_amplicon_trimmed2.fasta")
fasta_file3 <- readLines("~/Documents/DNA_reference_db_construction/refdb/final_library/trimming/right_amplicon_trimmed3.fasta")
fasta_file4 <- readLines("~/Documents/DNA_reference_db_construction/refdb/final_library/trimming/right_amplicon_trimmed4.fasta")

# merge into one
right_amplicon_trimmed_fasta <- c(fasta_file1,fasta_file2,fasta_file3,fasta_file4)
 #
right_amplicon_trimmed_fasta  <- data.frame(right_amplicon_trimmed_fasta)

write.table(right_amplicon_trimmed_fasta, file = "~/Documents/DNA_reference_db_construction/refdb/final_library/trimming/right_amplicon_trimmed.fasta", row.names = FALSE, col.names = FALSE, quote = FALSE)

```
#long amplicon
```{bash, engine.opts='-l'}

output_dir=$(pwd)
ampliconGroup=long
cutadapt -g GCHCCHGAYATRGCHTTYCC...TGRTTYTTYGGNCAYCCHGA -e 0.2 --revcomp -m 200 -M 550 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short1.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long1.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed1.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed1.fasta \
    ${output_dir}/all_db.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt.log

 # relax trim on too long 
cutadapt -g GCHCCHGAYATRGCHTTYCC...TGRTTYTTYGGNCAYCCHGA -e 0.32 --revcomp -m 200 -M 550 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short2.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long2.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed2.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed2.fasta \
    ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long1.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt2.log

 # relax trim on untrimmed    
cutadapt -g GCHCCHGAYATRGCHTTYCC...TGRTTYTTYGGNCAYCCHGA -e 0.35 --revcomp -m 200 -M 550 --max-n 0.05 --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short3.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long3.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed3.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed3.fasta \
    ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed1.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt3.log
    
cat ${output_dir}/trimming/long_amplicon_trimmed1.fasta ${output_dir}/trimming/long_amplicon_trimmed2.fasta ${output_dir}/trimming/long_amplicon_trimmed3.fasta ${output_dir}/trimming/long_amplicon_trimmed4.fasta > ${output_dir}/long_amplicon_trimmed.fasta
    
```    

# Join trimmed outputs back together into a single fasta file
```{r}
# join trimmed outputs
fasta_file1 <- readLines("/mnt/galaxy/home/mecarew/Documents/DNA_reference_db_construction/refdb/final_library/trimming/long_amplicon_trimmed1.fasta")
fasta_file2 <- readLines("/mnt/galaxy/home/mecarew/Documents/DNA_reference_db_construction/refdb/final_library/trimming/long_amplicon_trimmed2.fasta")
fasta_file3 <- readLines("~/Documents/DNA_reference_db_construction/refdb/final_library/trimming/long_amplicon_trimmed3.fasta")
long_amplicon_trimmed_fasta <- c(fasta_file1,fasta_file2,fasta_file3)

long_amplicon_trimmed_fasta  <- data.frame(long_amplicon_trimmed_fasta)

write.table(long_amplicon_trimmed_fasta, file = "~/Documents/DNA_reference_db_construction/refdb/final_library/trimming/long_amplicon_trimmed.fasta", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

#left amplicon 
```{bash, engine.opts='-l'}
ampliconGroup=left

cutadapt -g CCNGAYATRGCNTTYCC...CCNGTNYTNGCNGGNGCNATYACH -e 0.2 --revcomp -m 200 -M 450 --max-n 0.05  --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short1.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long1.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed1.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed1.fasta \
    ${output_dir}/all_db.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt.log

 # relax trim on too long 
cutadapt -g CCNGAYATRGCNTTYCC...CCNGTNYTNGCNGGNGCNATYACH -e 0.33 --revcomp -m 200 -M 450 --max-n 0.05   --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short2.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long2.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed2.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed2.fasta \
    ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long1.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt2.log

 # relax trim on untrimmed    
cutadapt -g CCNGAYATRGCNTTYCC...CCNGTNYTNGCNGGNGCNATYACH -e 0.37 --revcomp -m 200 -M 450 --max-n 0.05   --too-short-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_short3.fasta \
    --too-long-output ${output_dir}/trimming/${ampliconGroup}_amplicon_too_long3.fasta \
    --untrimmed-output ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed3.fasta \
    -o ${output_dir}/trimming/${ampliconGroup}_amplicon_trimmed3.fasta \
    ${output_dir}/trimming/${ampliconGroup}_amplicon_untrimmed1.fasta \
    > ${output_dir}/trimming/${ampliconGroup}_amplicon_cutadapt3.log

cat ${output_dir}/trimming/left_amplicon_trimmed1.fasta ${output_dir}/trimming/left_amplicon_trimmed2.fasta ${output_dir}/trimming/left_amplicon_trimmed3.fasta ${output_dir}/trimming/left_amplicon_trimmed4.fasta > ${output_dir}/left_amplicon_trimmed.fasta
  
     
     
```      

# Join trimmed outputs back together into a single fasta file
```{r}
# join trimmed outputs
fasta_file1 <- readLines("/mnt/galaxy/home/mecarew/Documents/DNA_reference_db_construction/refdb/final_library/trimming/left_amplicon_trimmed1.fasta")
fasta_file2 <- readLines("/mnt/galaxy/home/mecarew/Documents/DNA_reference_db_construction/refdb/final_library/trimming/left_amplicon_trimmed2.fasta")
fasta_file3 <- readLines("~/Documents/DNA_reference_db_construction/refdb/final_library/trimming/left_amplicon_trimmed3.fasta")
left_amplicon_trimmed_fasta <- c(fasta_file1,fasta_file2,fasta_file3)

left_amplicon_trimmed_fasta  <- data.frame(left_amplicon_trimmed_fasta)

write.table(left_amplicon_trimmed_fasta, file = "~/Documents/DNA_reference_db_construction/refdb/final_library/trimming/left_amplicon_trimmed.fasta", row.names = FALSE, col.names = FALSE, quote = FALSE)
```


##Needs more work (this is a library that would be used for DNA barcoding only as it include the part of the DNA barcode not used in metabarcoding) 
#LCO amplicon
ampliconGroup=LCO
cutadapt \
    --cores 4 \
    -g CAYAARGAYATTGG...GGNGGRTANACNGTTCANCC \
    -e 0.35 \
    --revcomp \
    -m 200 \
    -M 550 \
    --max-n 0.05 \
    --too-short-output ~/git/DNA_reference_db_construction/refdb/final_library/trimming/${ampliconGroup}_amplicon_too_short.fasta \
    --too-long-output ~/git/DNA_reference_db_construction/refdb/final_library/trimming/${ampliconGroup}_amplicon_too_long.fasta \
    --untrimmed-output ~/git/DNA_reference_db_construction/refdb/final_library/trimming/${ampliconGroup}_amplicon_untrimmed.fasta \
    -o ~/git/DNA_reference_db_construction/refdb/final_library/trimming/${ampliconGroup}_amplicon_trimmed.fasta \
    ~/git/DNA_reference_db_construction/refdb/final_library/all_db.fasta \
    > ~/git/DNA_reference_db_construction/refdb/final_library/trimming/${ampliconGroup}_amplicon_cutadapt.log



#Step 4: Merge sequences back together with taxonomic information.
# short amplicon (not need as completed with the writing script below)
```{r}
# read in trimmed short fasta
#short_fasta <- bioseq::read_fasta(file="~/git/DNA_reference_db_construction/refdb/final_library/trimming/short_amplicon_trimmed.fasta")
```

```{r}
# check format
#short_fasta %>% tail
```

```{r}
# check number of sequences
#length(short_fasta)
```

```{r}
# # Some sequences are now the reverse complement these need to be corrected
# rc_ids <- names(short_fasta)[! names(short_fasta) %in% rownames(all_db)]
# length(rc_ids)
# head(rc_ids)
```

```{r}
# # Check sequences in reverse compliment (rc)
# short_fasta["1927 rc"] %>% as.character
# short_fasta["1927 rc"] %>% seq_reverse %>% seq_complement %>% as.character
# all_db["1927", "sequence", drop=TRUE] %>% as.character
```

```{r}
# remove 'rc' from sequences
#short_db <- all_db[str_remove(names(short_fasta), " rc$"),]
#short_db$sequence <- short_fasta
```

```{r}
#table(is.na(short_db$id))
```

# Set parameters for writing files into library file format for the bioinformatic pipeline. 
The output flies will include a text file containing the taxonomic information ant a fasta file with the sequence data.
```{r}
# Write files
# Change from default behaviour (refdb::refdb_export_mothur and bioseq::write_fasta)
# - include taxonomy labels, e.g. k__Eukaryota;p__Arthropoda;c__Insecta
# - only use taxonmy levels: k, p, c, o, f, g, s
# - single line break between sequences and no block size
# Also add n_parts argument to split writing fasta file in multiple parts to avoid memory issues

write_db_files <- function (x, path_prefix, n_parts) {
    # Which taxonomy levels to include in output and the label names:
    tax_levels <- c("k" ="superkingdom", "p"="phylum", "c"="class", "o"="order", "f"="family",
                    "g"="genus", "s"="species")
    refdb:::check_fields(x, what = c("sequence", "taxonomy", "id"))
    # col_tax <- attributes(x)$refdb_fields$taxonomy
    col_tax <- tax_levels
    col_id <- attributes(x)$refdb_fields$id
    col_seq <- attributes(x)$refdb_fields$sequence
    labs <- x[[col_id]]
    labs <- stringr::str_replace_all(labs, "[:blank:]", "_")
    seqs <- x[[col_seq]]
    names(seqs) <- labs
    # Change NA values to ""
    x[, col_tax][is.na(x[, col_tax])] <- ""
    # Add labels names
    tax <- apply(x[, col_tax], 1, function(x) paste0(names(tax_levels), "__", x, collapse = ";"))
    tax <- paste0(labs, "\t", tax, ";")
    file_fas <- paste0(path_prefix, ".fasta")
    file_txt <- paste0(path_prefix, ".txt")
    if (n_parts <= 1) {
      write_fasta(seqs, file = file_fas)
    } else {
      write_fasta_parts(seqs, file = file_fas, n_parts=n_parts)
    }
    
    readr::write_lines(tax, file = file_txt)
}

write_fasta_parts <- function(x, file, n_parts) {
  parts <- as.numeric(cut(seq_along(x), n_parts))
  for (i in 1:n_parts) {
    if (i == 1) {
      write_fasta(x[parts == i], file = file,
                  line_length = 80, block_length=Inf, append=FALSE)
    } else {
      write_fasta(x[parts == i], file = file,
                  line_length = 80, block_length=Inf, append=TRUE)
    }
  }
}

write_fasta <- function (x, file, append = FALSE, line_length = 80, block_length = Inf) 
{
    x_nchar <- stringr::str_length(x)
    x_is_na <- is.na(x)
    if (line_length == Inf) {
        line_length <- max(x_nchar, na.rm = TRUE)
    }
    if (block_length == Inf) {
        block_length <- line_length
    }
    blocks_by_line <- line_length/block_length
    if (block_length > line_length) {
        stop("The length of blocks cannot be higher than the length of lines")
    }
    if (line_length%%block_length > 0L) {
        stop("The length of lines must be a multiple of the length of blocks")
    }
    if (any(x_is_na | x_nchar == 0L)) {
        input_len <- length(x)
        x <- x[!is.na(x)]
        output_len <- length(x)
        warning("Found ", input_len - output_len, " NA and/or empty sequences. They were not exported since ", 
            "the FASTA format does not support missing values.")
    }
    x <- as.character(x)
    x <- vapply(x, function(x) {
        x_len <- stringr::str_length(x)
        blocks <- seq(1, x_len, by = block_length)
        res <- stringr::str_sub(x, blocks, blocks - 1 + block_length)
        separator <- rep(" ", length(res))
        if (x_len > line_length) {
            separator[seq(blocks_by_line, blocks_by_line * (x_len%/%line_length), 
                blocks_by_line)] <- "\n"
        }
        separator[length(separator)] <- ""
        res <- stringr::str_c(res, separator, collapse = "")
        res
    }, vector("character", 1))
    fas <- stringr::str_c(">", names(x), sep = "")
    fas <- stringr::str_c(fas, x, sep = "\n", collapse = "\n") # Changed collapse to one line break
    readr::write_file(fas, file, append = append)
    readr::write_file("\n", file, append = TRUE) # Add one additional line break at the end
}
```

# Write all - set directories
```{r}
# Output prefix name and input file path
db_files <- c(
  "short_db_2023-07" = "~/git/DNA_barcode_reference_database_construction/final_library/short_amplicon_trimmed.fasta",
  "long_db_2023-07" = "~/git/DNA_barcode_reference_database_construction/final_library/long_amplicon_trimmed.fasta",
  "left_db_2023-07" = "~/git/DNA_barcode_reference_database_construction/final_library/left_amplicon_trimmed.fasta",
  "right_db_2023-07" = "~/git/DNA_barcode_reference_database_construction/final_library/right_amplicon_trimmed.fasta")
```

# Write all - write data as .txt files for taxonomic information and .fasta files with the sequence data
```{r}
output_dir <- "~/git/DNA_barcode_reference_database_construction/final_library"
for (i in seq_along(db_files)) {
  message("Processing ", names(db_files[i]))
  fasta <- bioseq::read_fasta(file=db_files[i])
  db <- all_db[str_remove(names(fasta), " rc$"),]
  db$sequence <- fasta
  db <- db %>% filter(! is.na(id))
  message("Writing files for ", names(db_files[i]))
  write_db_files(db, path_prefix=file.path(output_dir, names(db_files[i])), n_parts = 10)
}
```
